

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Workflows &mdash; seq2science  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Using the results" href="results.html" />
    <link rel="prev" title="Filling out the samples.tsv and config.yaml" href="fillingout.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="home.html">Welcome</a></li>
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="fillingout.html">Filling out the samples.tsv and config.yaml</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Workflows</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#downloading-fastqs">Downloading fastqs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pipeline-steps">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#download-sra">Download SRA</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fastq-dump">fastq-dump</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#best-practices">Best practices</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#filetype-extensions-and-paired-end-suffix">Filetype extensions and paired-end suffix</a></li>
<li class="toctree-l4"><a class="reference internal" href="#downloading-with-ascp">Downloading with ascp</a></li>
<li class="toctree-l4"><a class="reference internal" href="#downloading-many-samples">Downloading many samples</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fastq-dump-options">Fastq-dump options</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#alignment">Alignment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#automated-trimming">Automated trimming</a></li>
<li class="toctree-l4"><a class="reference internal" href="#alignment-sorting">Alignment &amp; Sorting</a></li>
<li class="toctree-l4"><a class="reference internal" href="#mark-duplicates">Mark duplicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#samtools-index">Samtools index</a></li>
<li class="toctree-l4"><a class="reference internal" href="#quality-report">Quality report</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trackhub">Trackhub</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id3">Best practices</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#automated-downloading-of-fastq">(Automated) downloading of fastq</a></li>
<li class="toctree-l4"><a class="reference internal" href="#working-with-local-files">Working with local files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#choosing-an-appropriate-aligner">Choosing an appropriate aligner</a></li>
<li class="toctree-l4"><a class="reference internal" href="#bam-mapping-quality-filtering">BAM mapping quality filtering</a></li>
<li class="toctree-l4"><a class="reference internal" href="#cram-support">Cram support</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#atac-seq">ATAC-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id6">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#peak-calling">Peak calling</a></li>
<li class="toctree-l4"><a class="reference internal" href="#replicates">Replicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id10">Quality report</a></li>
<li class="toctree-l4"><a class="reference internal" href="#count-table">count table</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id12">Trackhub</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id14">Best practices</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id15">(Automated) downloading of fastq</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id17">Alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#macs2-and-shift-with-paired-end-data">MACS2 and â€“shift with paired-end data</a></li>
<li class="toctree-l4"><a class="reference internal" href="#irreproducible-discovery-rate">Irreproducible Discovery Rate</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#rna-seq">RNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id19">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#generate-count-matrices">Generate count matrices</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id20">Trackhub</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-gene-expression-analysis">Differential gene expression analysis</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#contrast-designs">Contrast designs</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#conditions">Conditions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#batches">Batches</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#differential-gene-expression-analysis-output">Differential gene expression analysis output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#chip-seq">ChIP-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id24">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id25">Peak calling</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id34">Replicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id36">Quality report</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id39">count table</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id40">Trackhub</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id42">Best practices</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id43">(Automated) downloading of fastq</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id45">Alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id47">Irreproducible Discovery Rate</a></li>
<li class="toctree-l4"><a class="reference internal" href="#broad-peaks">Broad peaks</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#scatac-seq">scATAC-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id49">Pipeline steps</a></li>
<li class="toctree-l3"><a class="reference internal" href="#wip-optional-generate-an-annoation-file-which-cell-was-facsed-sorted-where-on-your-plate">WIP/ Optional, generate an annoation file which cell was FACsed sorted where on your plate</a></li>
<li class="toctree-l3"><a class="reference internal" href="#generate-and-bin-a-snap-object-for-downstream-analysis">Generate and bin a SNAP_Object for downstream analysis</a></li>
<li class="toctree-l3"><a class="reference internal" href="#how-to-get-the-pipeline-started">How to get the pipeline started?</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#filling-out-the-samples-tsv">2.6.1 Filling out the samples.tsv</a></li>
<li class="toctree-l4"><a class="reference internal" href="#filling-out-the-config-yaml">2.6.2 Filling out the config.yaml</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Workflows</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/workflows.md.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="workflows">
<h1>Workflows<a class="headerlink" href="#workflows" title="Permalink to this headline">Â¶</a></h1>
<div class="section" id="downloading-fastqs">
<h2>Downloading fastqs<a class="headerlink" href="#downloading-fastqs" title="Permalink to this headline">Â¶</a></h2>
<p>Downloading public data in bulk from the NCBI &amp; ENA databases has never been easier! See our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/tree/master/workflows/download_fastq">download fastq</a> workflow.</p>
<div class="section" id="pipeline-steps">
<h3>Pipeline steps<a class="headerlink" href="#pipeline-steps" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="download-sra">
<h4>Download SRA<a class="headerlink" href="#download-sra" title="Permalink to this headline">Â¶</a></h4>
<p>Most databases do not store fastqâ€™s, but they all store the raw data. For this reason for each sample submitted the raw data is downloaded. To convert this data to a fastq it has to be <em>dumped</em>.</p>
</div>
<div class="section" id="fastq-dump">
<h4>fastq-dump<a class="headerlink" href="#fastq-dump" title="Permalink to this headline">Â¶</a></h4>
<p>The second step is translating (dumping) the raw data to a fastq file.</p>
</div>
</div>
<div class="section" id="best-practices">
<h3>Best practices<a class="headerlink" href="#best-practices" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="filetype-extensions-and-paired-end-suffix">
<h4>Filetype extensions and paired-end suffix<a class="headerlink" href="#filetype-extensions-and-paired-end-suffix" title="Permalink to this headline">Â¶</a></h4>
<p>The pipeline will save fastq files in the <em>fastq_dir</em> directory, which is located in the <em>result_dir</em> directory by default. Fastqs already present in this directory will not be downloaded (again).</p>
<p>People and tools have different preferences for storing their data. One of these differences is how to name the fastq.gz files. Some people/tools prefer <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code>, while others prefer <code class="docutils literal notranslate"><span class="pre">fq.gz</span></code>. By default the pipeline names files with the fastq.gz extension. However if you prefer to change this you can set the variable <em>fqsuffix</em> accordingly.</p>
<p>The same goes for paired-end suffix conventions. Some people prefer for instance <code class="docutils literal notranslate"><span class="pre">sample_pass_1</span></code>, whilst others prefer <code class="docutils literal notranslate"><span class="pre">sample_R1</span></code> (the default). You can use your preferred suffix by setting variables <em>fqext1</em> and <em>fqext2</em>.
Additionally, the pipeline will search the <em>fastq_dir</em> directory for paired-ended files with unrecognized fqexts and attempt to parse them for you. As long as the suffixes are lexicographically ordered (R1 &gt; R2, pass1 &gt; pass2), you can ignore the <em>fqext</em> parameters.</p>
<p>In the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">fastq_dir: ./my_first_fastq_dir</span>

fqsuffix: fq
fqext1: R1
fqext2: R2
</pre></div>
</div>
<p>or on the command line:</p>
<p><code class="docutils literal notranslate"><span class="pre">snakemake</span> <span class="pre">--config</span> <span class="pre">fastq_dir=./my_first_fastq_dir</span> <span class="pre">fsuffix=fq</span> <span class="pre">fqext1=R1</span> <span class="pre">fqext2=R2</span></code></p>
</div>
<div class="section" id="downloading-with-ascp">
<h4>Downloading with ascp<a class="headerlink" href="#downloading-with-ascp" title="Permalink to this headline">Â¶</a></h4>
<p>ascp is a downloading protocol that allows for (sometimes much) faster speeds. If you have ascp installed you can make the workflow download through this protocol. You can specify the path to the binary in the config.yaml by key ascp_path and the ascp key in the config by key ascp_key. Or you could simply append the config on the command line, e.g.;</p>
<p><code class="docutils literal notranslate"><span class="pre">snakemake</span> <span class="pre">--config</span> <span class="pre">ascp_key=$HOME/.aspera/connect/etc/asperaweb_id_dsa.openssh</span> <span class="pre">ascp_path=$HOME/.aspera/connect/bin/ascp</span></code></p>
<p>See this <a class="reference external" href="https://gist.github.com/mfansler/71f09c8b6c9a95ec4e759a8ffc488be3">gist</a> for easy installation of ascp.</p>
</div>
<div class="section" id="downloading-many-samples">
<h4>Downloading many samples<a class="headerlink" href="#downloading-many-samples" title="Permalink to this headline">Â¶</a></h4>
<p>Before the pipeline starts downloading it will first determine whether the samples are paired-end or single-end. This is done by doing two requests per sample with the ncbi tools. Unregistered accounts can only make 3 requests per second. This means that initialization takes +/- 0.66s per sample. If you are an extremely impatient person like me, you can up these numbers a with a <a class="reference external" href="https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/">ncbi api key</a> (their default one allows 10 request p/s). You can set the variables <em>ncbi_key</em> and <em>ncbi_requests</em> accordingly in the config.yaml, or on the command line, e.g.;</p>
<p><code class="docutils literal notranslate"><span class="pre">snakemake</span> <span class="pre">--config</span> <span class="pre">ncbi_key=123456789</span> <span class="pre">ncbi_requests=10</span></code></p>
<p>The results of these lookups get cached, so when re-running parts of the workflow with the same samples no lookup online is required.</p>
</div>
<div class="section" id="fastq-dump-options">
<h4>Fastq-dump options<a class="headerlink" href="#fastq-dump-options" title="Permalink to this headline">Â¶</a></h4>
<p>We made a selection of <a class="reference external" href="https://edwards.sdsu.edu/research/fastq-dump/">reasonable parameters</a> for the dumping of SRAs to fastq. You can change these settings by changing the variable <em>splot</em> for single-end dumping, or the variable <em>split</em> for paired-end dumping, in the config.yaml, or on the command line.</p>
</div>
</div>
</div>
<div class="section" id="alignment">
<h2>Alignment<a class="headerlink" href="#alignment" title="Permalink to this headline">Â¶</a></h2>
<p>Aligning samples has never been easier! See our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/tree/master/workflows/alignment">alignment</a> workflow.</p>
<div class="section" id="id2">
<h3>Pipeline steps<a class="headerlink" href="#id2" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="automated-trimming">
<h4>Automated trimming<a class="headerlink" href="#automated-trimming" title="Permalink to this headline">Â¶</a></h4>
<p>The pipeline starts by trimming the reads with <a class="reference external" href="https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md">trim galore!</a>. Trim galore automatically first trims the low quality 3â€™ ends of reads, and removes short reads. After the quality trimming trim galore automatically detects which adapter was used, and trims it. The parameters of trim galore! for the pipeline can be set in the configuration by variable <em>trim_galore</em>.</p>
</div>
<div class="section" id="alignment-sorting">
<h4>Alignment &amp; Sorting<a class="headerlink" href="#alignment-sorting" title="Permalink to this headline">Â¶</a></h4>
<p>After trimming the reads are aligned against an assembly. Currently we support <em>bowtie2</em>, <em>bwa</em>, <em>hisat2</em> and <em>STAR</em> as aligners. Choosing which aligner is as easy as setting the <em>aligner</em> variable in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>, for example: <code class="docutils literal notranslate"><span class="pre">aligner:</span> <span class="pre">bwa</span></code>. Sensible defaults have been set for every aligner, but can be overwritten for either (or both) the indexing and alignment by specifying them in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code>:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">aligner:</span>
<span class="c">  star:</span>
<span class="c">    index: &#39;--limitGenomeGenerateRAM 55000000000`  # give STAR 55 GB RAM during indexing (default is 31 GB)</span>
<span class="c">    align: &#39;--readFilesCommand gunzip -c&#39;          # unzips fastqs before aligning</span>
</pre></div>
</div>
<p>The pipeline will check if the assembly you specified is present in the <em>genome_dir</em>, and otherwise will download it for you through <a class="reference external" href="https://github.com/vanheeringen-lab/genomepy">genomepy</a>. All these aligners require an index to be formed first for each assembly, but donâ€™t worry, the pipeline does this for you.</p>
<p>The outputted alignment.bam is immediately sorted by either samtools or sambamba (<em>bam_sorter</em>) either in <em>queryname</em> or <em>coordinate</em> (default) order. Take a look at our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/3.-Alignment#choosing-an-appropriate-aligner">Choosing an aligner</a> section for tips which aligner to use.</p>
</div>
<div class="section" id="mark-duplicates">
<h4>Mark duplicates<a class="headerlink" href="#mark-duplicates" title="Permalink to this headline">Â¶</a></h4>
<p>After aligning &amp; sorting the bam duplicate reads are being â€˜markedâ€™ by picard Markduplicates. You can change the call by setting <em>markduplicates</em> in config.yaml.</p>
</div>
<div class="section" id="samtools-index">
<h4>Samtools index<a class="headerlink" href="#samtools-index" title="Permalink to this headline">Â¶</a></h4>
<p>Many downstream tools require an index of the deduplicated bam. The pipeline automatically generates these for you.</p>
</div>
<div class="section" id="quality-report">
<h4>Quality report<a class="headerlink" href="#quality-report" title="Permalink to this headline">Â¶</a></h4>
<p>Since the pipeline does a lot of steps automatically, it is very useful to have a quality report of samples. Along the way different quality control steps are taken, and are outputted in a single <a class="reference external" href="https://multiqc.info/">multiqc report</a>. Make sure to always check the report, and see our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/9.-Interpreting-the-quality-report">explanation</a> on how to interpret the different scores!</p>
</div>
<div class="section" id="trackhub">
<h4>Trackhub<a class="headerlink" href="#trackhub" title="Permalink to this headline">Â¶</a></h4>
<p>A trackhub can be generated for this workflow. See the <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/Trackhub">Trackhub page</a> for more information!</p>
</div>
</div>
<div class="section" id="id3">
<h3>Best practices<a class="headerlink" href="#id3" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="automated-downloading-of-fastq">
<h4>(Automated) downloading of fastq<a class="headerlink" href="#automated-downloading-of-fastq" title="Permalink to this headline">Â¶</a></h4>
<p>Alignment can be done with local fastq files, downloaded files, or a mix of those. Make sure to take a look at our downloading fastq <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/2.-Downloading-samples">best practices</a> when making use of the downloading functionality.</p>
</div>
<div class="section" id="working-with-local-files">
<h4>Working with local files<a class="headerlink" href="#working-with-local-files" title="Permalink to this headline">Â¶</a></h4>
<p>See our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/1.-Getting-started#where-does-the-workflow-store-results-and-looks-for-starting-points">Getting started</a> about working with local files.</p>
</div>
<div class="section" id="choosing-an-appropriate-aligner">
<h4>Choosing an appropriate aligner<a class="headerlink" href="#choosing-an-appropriate-aligner" title="Permalink to this headline">Â¶</a></h4>
<p>The RNA-seq workflow supports STAR and Salmon. By default, STAR is selected.</p>
<ul class="simple">
<li><p>STAR required large amounts of memory to index the genome, but is <a class="reference external" href="https://link.springer.com/article/10.1186/1471-2105-14-91">the more accurate</a> of the two.</p></li>
<li><p>Salmon allows for fast and light-weight transcript quantification. The accuracy can be improved by using a decoy-aware index. This can be achieved by setting <em>decoy_aware_index</em> to True in the config.yaml. However, the generation of the decoy transcripts requires an even larger amounts of memory compared to STAR.</p></li>
</ul>
<p>ATAC-seq workflows often use BWA or Bowtie2. By default, BWA is selected.</p>
</div>
<div class="section" id="bam-mapping-quality-filtering">
<h4>BAM mapping quality filtering<a class="headerlink" href="#bam-mapping-quality-filtering" title="Permalink to this headline">Â¶</a></h4>
<p>All aligners pass a mapping quality (MAPQ) score to the reads. This score reflects the certainty that a read belongs to a certain genomic position. This score can then be used by <a class="reference external" href="https://deeptools.readthedocs.io/en/develop/">deeptools</a> to only keep reads of which the aligner is certain where it aligns, and to limit multimappers. You can set the minimum mapq score by using the configuration variable <code class="docutils literal notranslate"><span class="pre">min_mapping_quality</span></code>. Unfortunately, most aligners use a different scoring system. This problem has been nicely summarized <a class="reference external" href="https://sequencing.qcfail.com/articles/mapq-values-are-really-useful-but-their-implementation-is-a-mess/">here</a>.</p>
<p>For the default aligners (BWA and STAR), a minimal quality score has been set to select uniquely mapped reads only. For other aligners, we suggest the following scores to filter for uniquely mapped reads:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Aligner</p></th>
<th class="head"><p>min. MAPQ scores for uniquely mapped reads</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Bowtie2</p></td>
<td><p>41</p></td>
</tr>
<tr class="row-odd"><td><p>BWA</p></td>
<td><p>30</p></td>
</tr>
<tr class="row-even"><td><p>Hisat2</p></td>
<td><p>44</p></td>
</tr>
<tr class="row-odd"><td><p>STAR</p></td>
<td><p>255</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="cram-support">
<h4>Cram support<a class="headerlink" href="#cram-support" title="Permalink to this headline">Â¶</a></h4>
<p>For the edge-case where you want to work with cram files instead of bams, there is an option <code class="docutils literal notranslate"><span class="pre">cram_no_bam</span></code> which will convert your bam files to cram (saves around 60% storage).</p>
</div>
</div>
</div>
<div class="section" id="atac-seq">
<h2>ATAC-seq<a class="headerlink" href="#atac-seq" title="Permalink to this headline">Â¶</a></h2>
<p>Running an ATAC-seq pipeline has never been easier! See our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/tree/master/workflows/atac_seq">ATAC-seq</a> workflow.</p>
<div class="section" id="id6">
<h3>Pipeline steps<a class="headerlink" href="#id6" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="peak-calling">
<h4>Peak calling<a class="headerlink" href="#peak-calling" title="Permalink to this headline">Â¶</a></h4>
<div class="section" id="macs2">
<h5>MACS2<a class="headerlink" href="#macs2" title="Permalink to this headline">Â¶</a></h5>
<p><a class="reference external" href="https://github.com/taoliu/MACS">MACS2</a> is the successor of MACS, created by <a class="reference external" href="https://github.com/taoliu">taoliu</a> (also one of the co-authors of hmmratac). MACS2 generates a â€˜<em>pileup</em>â€™ of all the reads. You can imagine this pileup as laying all reads on top of each other on the genome, and counting how high your pile gets for each basepair. MACS2 then models â€˜<em>background</em>â€™ values, based on the total length of all your reads and the genome length, to a <a class="reference external" href="https://en.wikipedia.org/wiki/Poisson_distribution">Poisson distribution</a> and decides whether your experimental pileup is significant compared to the poisson distribution. MACS2 is one of the (if not the) most used peak caller.</p>
<p>For the calculation of peaks MACS2 requires that a genome size is being passed as one of its arguments. For the more common assemblies (e.g. mm10 and human) these numbers can be found online. However we found googling these numbers quite the hassle, so the pipeline automatically estimates this number (we calculate the genome size as <strong>the number of unique kmers of the average read length</strong>).</p>
</div>
<div class="section" id="genrich">
<h5>Genrich<a class="headerlink" href="#genrich" title="Permalink to this headline">Â¶</a></h5>
<p><a class="reference external" href="https://github.com/jsh58/Genrich">Genrich</a> is the spiritual successor of MACS2, created by <a class="reference external" href="https://github.com/jsh58">John M. Gaspar</a>. Just like MACS2 is generates a â€˜<em>pileupâ€™</em>. However the author of genrich realized that the distribution of pileup never follows a poisson distribution. Genrich then uses a <a class="reference external" href="https://en.wikipedia.org/wiki/Log-normal_distribution">log-normal distribution</a> to model the background. Genrich has many nice features, but is a relatively new and untested, and is not published yet.</p>
</div>
<div class="section" id="hmmratac-not-fully-supported">
<h5>HMMRATAC  (Not fully supported!)<a class="headerlink" href="#hmmratac-not-fully-supported" title="Permalink to this headline">Â¶</a></h5>
<p><a class="reference external" href="https://github.com/LiuLabUB/HMMRATAC">HMMRATAC</a> is a peak caller specifically made for ATAC-seq data, created by <a class="reference external" href="https://github.com/taoliu">taoliu</a>. HMMRATAC does not generate a so-called pileup. It however looks at the length distribution of reads. Since the tn5 transposase prefers to cut at nucleosome free regions in the DNA, we see that our reads are either short (at open DNA), or the length of 1 (or 2, 3, 4, etc.) nucleosomes long. HMMRATAC then classifies regions in the genome based on the length distribution of the reads in open chromatin, flanking regions, non-flanking regions and background.</p>
</div>
</div>
<div class="section" id="replicates">
<h4>Replicates<a class="headerlink" href="#replicates" title="Permalink to this headline">Â¶</a></h4>
<p>All peak callers are capable of combining biological replicates. See <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/4.-Replicate-handling">Replicate handling</a> for more information.</p>
</div>
<div class="section" id="id10">
<h4>Quality report<a class="headerlink" href="#id10" title="Permalink to this headline">Â¶</a></h4>
<p>Since the pipeline does a lot of steps automatically, it is very useful to have a quality report of samples. Along the way different quality control steps are taken, and are outputted in a single <a class="reference external" href="https://multiqc.info/">multiqc report</a> in the <code class="docutils literal notranslate"><span class="pre">qc</span></code> folder. Make sure to always check the report, and take a look at <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/3.0-Interpreting-the-MultiQC-report">interpreting the multiqc report</a>!</p>
</div>
<div class="section" id="count-table">
<h4>count table<a class="headerlink" href="#count-table" title="Permalink to this headline">Â¶</a></h4>
<p>A useful result the pipeline outputs is the â€˜count tableâ€™, located at {peak_caller}/count<em>table</em>{assembly}.bed. For every peak the peak caller found in your data, it outputs the number of reads for each sample that was under that peak.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>                sample1     sample2
FLLO01000001.1:5649-5849    75.00000    0.00000
FLLO01000001.1:7399-7599    47.00000    1.00000
</pre></div>
</div>
<p><strong>Note</strong> that this uses the number of raw reads under each peak, and does not use any kind of correction for the difference in total number of reads between samples!</p>
</div>
<div class="section" id="id12">
<h4>Trackhub<a class="headerlink" href="#id12" title="Permalink to this headline">Â¶</a></h4>
<p>A trackhub is be generated for this workflow. See the <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/3.1-Trackhub">Trackhub page</a> for more information!</p>
</div>
</div>
<div class="section" id="id14">
<h3>Best practices<a class="headerlink" href="#id14" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="id15">
<h4>(Automated) downloading of fastq<a class="headerlink" href="#id15" title="Permalink to this headline">Â¶</a></h4>
<p>ATAC-seq can be done with local fastq files, downloaded files, or a mix of those. Make sure to take a look at our downloading fastq <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/2.-Downloading-samples">best practices</a> when making use of the downloading functionality.</p>
</div>
<div class="section" id="id17">
<h4>Alignment<a class="headerlink" href="#id17" title="Permalink to this headline">Â¶</a></h4>
<p>Make sure to take a look at our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/3.-Alignment">alignment wiki</a>.</p>
</div>
<div class="section" id="macs2-and-shift-with-paired-end-data">
<h4>MACS2 and â€“shift with paired-end data<a class="headerlink" href="#macs2-and-shift-with-paired-end-data" title="Permalink to this headline">Â¶</a></h4>
<p>When trying to detect the cutting sites of Tn5, it is common to shift all your reads for instance 100 bp upstream, and extend them to 200 long. This makes the cutting site of a read fall in the middle of this 200 bp region. Unfortunately this is only possible in single-end mode of MACS, and makes MACS2 throw away all(!) of the paired-end mates, resulting in only half the amount of readsâ€¦</p>
<p>Fortunately for you, we made a script that deletes all the information that reads are paired-end after alignment from a bam. In this way we bypass the problem that shift is only possible for single-end reads. If you want to make use of this option then set the <em>macs2_keep_mates</em> variable in the configuration to True.</p>
</div>
<div class="section" id="irreproducible-discovery-rate">
<h4>Irreproducible Discovery Rate<a class="headerlink" href="#irreproducible-discovery-rate" title="Permalink to this headline">Â¶</a></h4>
<p>For idr to work properly, it needs a (large) portion of peaks that are actually not true peaks. Therefore we recommend to call peaks â€˜looselyâ€™. One way of doing this is setting the <a class="reference external" href="https://github.com/taoliu/MACS#-q--pvalue">peak threshold q-value</a> high for macs2 in the configuration.</p>
</div>
</div>
</div>
<div class="section" id="rna-seq">
<h2>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">Â¶</a></h2>
<p>Running an RNA-seq pipeline has never been easier! See our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/tree/master/workflows/rna_seq">RNA-seq</a> workflow.</p>
<div class="section" id="id19">
<h3>Pipeline steps<a class="headerlink" href="#id19" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="generate-count-matrices">
<h4>Generate count matrices<a class="headerlink" href="#generate-count-matrices" title="Permalink to this headline">Â¶</a></h4>
<p>In addition to aligning to the genome, <a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> can output a reads per gene table. These counts are required for differential expression analysis.</p>
<p>If alignment is no priority, either STAR or <a class="reference external" href="https://github.com/COMBINE-lab/salmon">Salmon</a> can be used to only quantify the transcripts. This option is used when <code class="docutils literal notranslate"><span class="pre">generate_trackhub</span></code> is set to <code class="docutils literal notranslate"><span class="pre">False</span></code> (the default for this workflow).</p>
</div>
<div class="section" id="id20">
<h4>Trackhub<a class="headerlink" href="#id20" title="Permalink to this headline">Â¶</a></h4>
<p>A trackhub can be generated for this workflow. See the <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/Trackhub">Trackhub page</a> for more information!</p>
</div>
<div class="section" id="differential-gene-expression-analysis">
<h4>Differential gene expression analysis<a class="headerlink" href="#differential-gene-expression-analysis" title="Permalink to this headline">Â¶</a></h4>
<p>Differential expression analysis can automatically be performed using <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">DESeq2</a>. Simply add one or more contrast design(s) in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file. Examples are given below. Additionally, the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> contains a commented-out set of examples.</p>
<p>Note: (additional) design contrasts can be added at any time. After completing the workflow, rerunning Snakemake with new contrasts will only perform these analyses.</p>
<div class="section" id="id22">
<h5>DESeq2<a class="headerlink" href="#id22" title="Permalink to this headline">Â¶</a></h5>
<p>DESeq2 automatically performs library bias correction when loading your data. Batch correction is performed if a batch effect is included in the design. After calculating differentially expressed genes, a multiple testing procedure is applied. By default this is the Benjamini-Hochberg procedure, although Independent Hypothesis Weighing can also be applied. Expression counts are log transformed (by default using the apeglm method). Finally, a list of all genes is saved to file, with analysis results for expressed genes.</p>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="contrast-designs">
<h3>Contrast designs<a class="headerlink" href="#contrast-designs" title="Permalink to this headline">Â¶</a></h3>
<p>The following section will guide you through the process of creating a DESeq2 contrast using only samples.tsv and the config.yaml.</p>
<p>A contrast contains a condition, which is the variable you wish to analyse, and which contains 2 or more states. Optionally, a contrast can contain a batch effect, which you want to correct for.</p>
<div class="section" id="conditions">
<h4>Conditions<a class="headerlink" href="#conditions" title="Permalink to this headline">Â¶</a></h4>
<p>In samples.tsv, add a column for every property of your samples.
Examples:</p>
<ul class="simple">
<li><p>a column named â€˜conditionsâ€™ with values â€˜controlâ€™ and â€˜treatedâ€™.</p></li>
<li><p>a column named â€˜conditionsâ€™ with values â€˜controlâ€™, â€˜treatmentAâ€™ and â€˜treatmentBâ€™.</p></li>
<li><p>a column named â€˜stagesâ€™ with values 1, 2 and 3.</p></li>
</ul>
<p>Next, in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> add one or more contrasts.</p>
<p>In order to compare two groups (e.a. control vs treated) in a column with only two groups, the contrast condition is the column name (<code class="docutils literal notranslate"><span class="pre">conditions</span></code>) (the next notation is allowed as well).</p>
<p>In order to compare two groups for a larger list (e.a. stage 1 vs stage 3, but not stage 2), the contrast condition is the column name, followed by the groups (<code class="docutils literal notranslate"><span class="pre">stages_3_1</span></code>).</p>
<p>To compare all groups against a reference (e.a. control vs both treatments A and B), the contrast condition is the column name, followed by the reference level (<code class="docutils literal notranslate"><span class="pre">condition_control</span></code>). Alternatively, group â€˜allâ€™ may also be added (<code class="docutils literal notranslate"><span class="pre">condition_all_control</span></code>).</p>
</div>
<div class="section" id="batches">
<h4>Batches<a class="headerlink" href="#batches" title="Permalink to this headline">Â¶</a></h4>
<p>If your data has batch effects, these can be accounted for. Note that this can only be done if the batch effects are spread over different conditions.</p>
<p>Similar to conditions, add a column to samples.tsv and note the batch of each sample:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>sample</p></th>
<th class="head"><p>sequencing_month</p></th>
<th class="head"><p>conditions</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sample_1</p></td>
<td><p>jan</p></td>
<td><p>control</p></td>
</tr>
<tr class="row-odd"><td><p>sample_2</p></td>
<td><p>feb</p></td>
<td><p>control</p></td>
</tr>
<tr class="row-even"><td><p>sample_3</p></td>
<td><p>jan</p></td>
<td><p>treatment</p></td>
</tr>
<tr class="row-odd"><td><p>sample_4</p></td>
<td><p>feb</p></td>
<td><p>treatment</p></td>
</tr>
</tbody>
</table>
<p>In this case, the sequencing_month may have caused a batch effect. Since the (potential) batch effect is spread over the test conditions (control vs treatment) it can be taken into account during DE analysis. The contrast design to do this would be <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_treatment_control</span></code></p>
<div class="section" id="more-examples">
<h5>More examples<a class="headerlink" href="#more-examples" title="Permalink to this headline">Â¶</a></h5>
<p>The example design <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_treatment_control</span></code> may also be written as <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions</span></code> or <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_all_control</span></code>.</p>
<p>For consistency with regular DESeq2 contrast designs, designs may start with <code class="docutils literal notranslate"><span class="pre">~</span></code>, example: <code class="docutils literal notranslate"><span class="pre">~</span> <span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_treatment_control</span></code>.</p>
<p>Our gremlins will carefully unpack all of your input and pass the unique designs to DESeq2.</p>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="differential-gene-expression-analysis-output">
<h3>Differential gene expression analysis output<a class="headerlink" href="#differential-gene-expression-analysis-output" title="Permalink to this headline">Â¶</a></h3>
<p>Each contrast will return a table with all genes found during alignment.</p>
<ul class="simple">
<li><p>The column <code class="docutils literal notranslate"><span class="pre">padj</span></code> contains the adjusted p-values after multiple testing. These should be used extract DE genes.</p></li>
<li><p>The column <code class="docutils literal notranslate"><span class="pre">log2FoldChange</span></code> shows the fold change of each gene between the two conditions. The reference condition is the one <em>last mentioned</em> in the contrast design, so use <code class="docutils literal notranslate"><span class="pre">conditions_treatment_control</span></code>. If you use <code class="docutils literal notranslate"><span class="pre">conditions_control_treatment</span></code> the fold change is <em>inverted</em>.</p></li>
<li><p>Several other columns were kept for sake of completion, such as column <code class="docutils literal notranslate"><span class="pre">pvalue</span></code>, which contains non-adjusted p-values and <em>should not be used</em>.</p></li>
</ul>
<p>Additional outputs include a blind clustering (which may reveal switched sample names), and per contrast an MA plot and a PCA plot (to accompany the differentially expressed genes table).</p>
</div>
</div>
<div class="section" id="chip-seq">
<h2>ChIP-seq<a class="headerlink" href="#chip-seq" title="Permalink to this headline">Â¶</a></h2>
<p>Running a CHiP-seq pipeline has never been easier! See our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/tree/master/workflows/chip_seq">CHiP-seq</a> workflow.</p>
<div class="section" id="id24">
<h3>Pipeline steps<a class="headerlink" href="#id24" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="id25">
<h4>Peak calling<a class="headerlink" href="#id25" title="Permalink to this headline">Â¶</a></h4>
<div class="section" id="id26">
<h5>MACS2<a class="headerlink" href="#id26" title="Permalink to this headline">Â¶</a></h5>
<p><a class="reference external" href="https://github.com/taoliu/MACS">MACS2</a> is the successor of MACS, created by <a class="reference external" href="https://github.com/taoliu">taoliu</a> (also one of the co-authors of hmmratac). MACS2 generates a â€˜<em>pileup</em>â€™ of all the reads. You can imagine this pileup as laying all reads on top of each other on the genome, and counting how high your pile gets for each basepair. MACS2 then models â€˜<em>background</em>â€™ values, based on the total length of all your reads and the genome length, to a <a class="reference external" href="https://en.wikipedia.org/wiki/Poisson_distribution">Poisson distribution</a> and decides whether your experimental pileup is significant compared to the poisson distribution. MACS2 is one of the (if not the) most used peak caller.</p>
<p>For the calculation of peaks MACS2 requires that a genome size is being passed as one of its arguments. For the more common assemblies (e.g. mm10 and human) these numbers can be found online. However we found googling these numbers quite the hassle, so the pipeline automatically estimates this number (we calculate the genome size as <strong>the number of unique kmers of the average read length</strong>).</p>
</div>
<div class="section" id="id30">
<h5>Genrich<a class="headerlink" href="#id30" title="Permalink to this headline">Â¶</a></h5>
<p><a class="reference external" href="https://github.com/jsh58/Genrich">Genrich</a> is the spiritual successor of MACS2, created by <a class="reference external" href="https://github.com/jsh58">John M. Gaspar</a>. Just like MACS2 is generates a â€˜<em>pileupâ€™</em>. However the author of genrich realized that the distribution of pileup never follows a poisson distribution. Genrich then uses a <a class="reference external" href="https://en.wikipedia.org/wiki/Log-normal_distribution">log-normal distribution</a> to model the background. Genrich has many nice features, but is a relatively new and untested, and is not published yet.</p>
</div>
</div>
<div class="section" id="id34">
<h4>Replicates<a class="headerlink" href="#id34" title="Permalink to this headline">Â¶</a></h4>
<p>All peak callers are capable of combining biological replicates. See <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/1.1-Filling-out-the-samples.tsv#atac--and-chip-seq">Replicate handling</a> for more information.</p>
</div>
<div class="section" id="id36">
<h4>Quality report<a class="headerlink" href="#id36" title="Permalink to this headline">Â¶</a></h4>
<p>Since the pipeline does a lot of steps automatically, it is very useful to have a quality report of samples. Along the way different quality control steps are taken, and are outputted in a single <a class="reference external" href="https://multiqc.info/">multiqc report</a> in the <code class="docutils literal notranslate"><span class="pre">qc</span></code> folder. Make sure to always check the report, and take a look at <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/3.0-Interpreting-the-MultiQC-report">interpreting the multiqc report</a>!</p>
</div>
<div class="section" id="id39">
<h4>count table<a class="headerlink" href="#id39" title="Permalink to this headline">Â¶</a></h4>
<p>A useful result the pipeline outputs is the â€˜count tableâ€™, located at {peak_caller}/count<em>table</em>{assembly}.bed. For every peak the peak caller found in your data, it outputs the number of reads for each sample that was under that peak.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>                sample1     sample2
FLLO01000001.1:5649-5849    75.00000    0.00000
FLLO01000001.1:7399-7599    47.00000    1.00000
</pre></div>
</div>
<p><strong>Note</strong> that this uses the number of raw reads under each peak, and does not use any kind of correction for the difference in total number of reads between samples!</p>
</div>
<div class="section" id="id40">
<h4>Trackhub<a class="headerlink" href="#id40" title="Permalink to this headline">Â¶</a></h4>
<p>A trackhub is be generated for this workflow. See the <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/3.1-Trackhub">Trackhub page</a> for more information!</p>
</div>
</div>
<div class="section" id="id42">
<h3>Best practices<a class="headerlink" href="#id42" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="id43">
<h4>(Automated) downloading of fastq<a class="headerlink" href="#id43" title="Permalink to this headline">Â¶</a></h4>
<p>ATAC-seq can be done with local fastq files, downloaded files, or a mix of those. Make sure to take a look at our downloading fastq <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/2.-Downloading-samples">best practices</a> when making use of the downloading functionality.</p>
</div>
<div class="section" id="id45">
<h4>Alignment<a class="headerlink" href="#id45" title="Permalink to this headline">Â¶</a></h4>
<p>Make sure to take a look at our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/3.-Alignment">alignment wiki</a>.</p>
</div>
<div class="section" id="id47">
<h4>Irreproducible Discovery Rate<a class="headerlink" href="#id47" title="Permalink to this headline">Â¶</a></h4>
<p>For idr to work properly, it needs a (large) portion of peaks that are actually not true peaks. Therefore we recommend to call peaks â€˜looselyâ€™. One way of doing this is setting the <a class="reference external" href="https://github.com/taoliu/MACS#-p--pvalue">peak threshold p-value</a> high for macs2 in the configuration.</p>
</div>
<div class="section" id="broad-peaks">
<h4>Broad peaks<a class="headerlink" href="#broad-peaks" title="Permalink to this headline">Â¶</a></h4>
<p>Some histone marks (such as H3K9me3 and H3K27me3) produce broader (wider) peaks. If you want macs2 to call broad peaks you need to add the <code class="docutils literal notranslate"><span class="pre">--broad</span></code> flag in the macs2 call in the config. Furthermore additional tweaking of the broadness of your peaks is possible by also adding the options: <code class="docutils literal notranslate"><span class="pre">--broad-cutoff</span></code> and especially <code class="docutils literal notranslate"><span class="pre">--max-gap</span></code> in the macs call in the config.
For example using H3K27me3 data you can use  <code class="docutils literal notranslate"><span class="pre">--broad-cutoff</span> <span class="pre">0.1</span></code>, and  <code class="docutils literal notranslate"><span class="pre">--max-gap</span> <span class="pre">10000</span></code> (10kb), this will merge peaks closer than 10kb together as shown bellow:</p>
<p>WIP: Genome browser image comparing peak tracks</p>
</div>
</div>
</div>
<div class="section" id="scatac-seq">
<h2>scATAC-seq<a class="headerlink" href="#scatac-seq" title="Permalink to this headline">Â¶</a></h2>
<p>Running a scATAC-seq pipeline has never been easier! See our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/tree/master/workflows/scATAC_seq">scATAC-seq</a> workflow.</p>
<p>Does your scATAC protocol generate single cell fastq files? And do they need to be mapped and analysed? Then this pipeline is for your! This pipeline takes single cell fastq files, performs extensive (plate based) QC steps, and generates a binnedc SNAP object that can be used for further downstream analysis following the tutorials on: <a class="reference external" href="https://github.com/r3fang/SnapATAC">https://github.com/r3fang/SnapATAC</a></p>
<p>If instead you have a large FASTQ file containing a cell specific ID in the header, it might possible to run this pipeline with some alterations. If you have instead a Cellranger object, it is relatively easy to directly generate a SNAPobject from those, look at the doucmentation on the SnapATAC github intead.</p>
<div class="section" id="id49">
<h3>Pipeline steps<a class="headerlink" href="#id49" title="Permalink to this headline">Â¶</a></h3>
<p>Concatinate samples to large FASTQ file and add cell ID to fastq header
Run Trim-Galore to remove adapters and perform QC testing of the fastq files
align using hisat2, bwa or star.</p>
</div>
<div class="section" id="wip-optional-generate-an-annoation-file-which-cell-was-facsed-sorted-where-on-your-plate">
<h3>WIP/ Optional, generate an annoation file which cell was FACsed sorted where on your plate<a class="headerlink" href="#wip-optional-generate-an-annoation-file-which-cell-was-facsed-sorted-where-on-your-plate" title="Permalink to this headline">Â¶</a></h3>
<p>Depending on how you generated your scATAC data, you potentially have a FACS sorting based protocol. If this is the case you can semi-automate the annotation the barcoded fastq files to each well (and therefore FACs sorted cell type).
In order to perform this annotation, in the folder XXX WIP XXX there will be a annotation.toml file.</p>
<p>Do here XXX WIP XXX to generate a barcode-celltype file that can be used in downstream annotation of the SNAP-object.</p>
</div>
<div class="section" id="generate-and-bin-a-snap-object-for-downstream-analysis">
<h3>Generate and bin a SNAP_Object for downstream analysis<a class="headerlink" href="#generate-and-bin-a-snap-object-for-downstream-analysis" title="Permalink to this headline">Â¶</a></h3>
<p>To find cellular heterogeneity before peak calling, SNAP-ATAC binns the genome. Using this for demensionality reduction of the data and finding cell clusters before peak calling. By default the pipeline includes bins of 5kB, but this is changable in the config.yaml file.</p>
</div>
<div class="section" id="how-to-get-the-pipeline-started">
<h3>How to get the pipeline started?<a class="headerlink" href="#how-to-get-the-pipeline-started" title="Permalink to this headline">Â¶</a></h3>
<div class="section" id="filling-out-the-samples-tsv">
<h4>2.6.1 Filling out the samples.tsv<a class="headerlink" href="#filling-out-the-samples-tsv" title="Permalink to this headline">Â¶</a></h4>
</div>
<div class="section" id="filling-out-the-config-yaml">
<h4>2.6.2 Filling out the config.yaml<a class="headerlink" href="#filling-out-the-config-yaml" title="Permalink to this headline">Â¶</a></h4>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="results.html" class="btn btn-neutral float-right" title="Using the results" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="fillingout.html" class="btn btn-neutral float-left" title="Filling out the samples.tsv and config.yaml" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>