

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Per rule explanation &mdash; seq2science  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="All configurable options" href="schemas.html" />
    <link rel="prev" title="Command line interface" href="cli.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Using the results</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="extensive_docs.html">Extensive docs</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Per rule explanation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#bam2cram">bam2cram</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bam-bigwig">bam_bigwig</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bam-stranded-bigwig">bam_stranded_bigwig</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bedgraph-bigwig">bedgraph_bigwig</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bedgraphish-to-bedgraph">bedgraphish_to_bedgraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="#blind-clustering">blind_clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bowtie2-align">bowtie2_align</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bowtie2-index">bowtie2_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bwa-index">bwa_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bwa-mem">bwa_mem</a></li>
<li class="toctree-l3"><a class="reference internal" href="#call-peak-genrich">call_peak_genrich</a></li>
<li class="toctree-l3"><a class="reference internal" href="#complement-blacklist">complement_blacklist</a></li>
<li class="toctree-l3"><a class="reference internal" href="#computematrix">computeMatrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="#counts-matrix">counts_matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="#create-snap-object">create_SNAP_object</a></li>
<li class="toctree-l3"><a class="reference internal" href="#create-bins-snap-object">create_bins_SNAP_object</a></li>
<li class="toctree-l3"><a class="reference internal" href="#cytoband">cytoband</a></li>
<li class="toctree-l3"><a class="reference internal" href="#decoy-transcripts">decoy_transcripts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#deseq2">deseq2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fastqc">fastqc</a></li>
<li class="toctree-l3"><a class="reference internal" href="#featurecounts">featureCounts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#gcpercent">gcPercent</a></li>
<li class="toctree-l3"><a class="reference internal" href="#genrich-pileup">genrich_pileup</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-genome">get_genome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#get-transcripts">get_transcripts</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hisat2-align">hisat2_align</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hisat2-index">hisat2_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hmmratac">hmmratac</a></li>
<li class="toctree-l3"><a class="reference internal" href="#hmmratac-genome-info">hmmratac_genome_info</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id2sra">id2sra</a></li>
<li class="toctree-l3"><a class="reference internal" href="#idr">idr</a></li>
<li class="toctree-l3"><a class="reference internal" href="#insert-size-metrics">insert_size_metrics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#keep-mates">keep_mates</a></li>
<li class="toctree-l3"><a class="reference internal" href="#linked-txome">linked_txome</a></li>
<li class="toctree-l3"><a class="reference internal" href="#macs2-callpeak">macs2_callpeak</a></li>
<li class="toctree-l3"><a class="reference internal" href="#macs-bdgcmp">macs_bdgcmp</a></li>
<li class="toctree-l3"><a class="reference internal" href="#macs-cmbreps">macs_cmbreps</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mark-duplicates">mark_duplicates</a></li>
<li class="toctree-l3"><a class="reference internal" href="#merge-replicates">merge_replicates</a></li>
<li class="toctree-l3"><a class="reference internal" href="#mt-nuc-ratio-calculator">mt_nuc_ratio_calculator</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multibamsummary">multiBamSummary</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc">multiqc</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc-filter-buttons">multiqc_filter_buttons</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc-header-info">multiqc_header_info</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc-rename-buttons">multiqc_rename_buttons</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc-samplesconfig">multiqc_samplesconfig</a></li>
<li class="toctree-l3"><a class="reference internal" href="#multiqc-schema">multiqc_schema</a></li>
<li class="toctree-l3"><a class="reference internal" href="#peak-bigpeak">peak_bigpeak</a></li>
<li class="toctree-l3"><a class="reference internal" href="#peak-union">peak_union</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plotcorrelation">plotCorrelation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plotfingerprint">plotFingerprint</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plotpca">plotPCA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#plotprofile">plotProfile</a></li>
<li class="toctree-l3"><a class="reference internal" href="#renamefastq-pe">renamefastq_PE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#salmon-decoy-aware-index">salmon_decoy_aware_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#salmon-index">salmon_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#salmon-quant">salmon_quant</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sambamba-sort">sambamba_sort</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-index">samtools_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-index-cram">samtools_index_cram</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-presort">samtools_presort</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-sort">samtools_sort</a></li>
<li class="toctree-l3"><a class="reference internal" href="#samtools-stats">samtools_stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="#setup-blacklist">setup_blacklist</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sieve-bam">sieve_bam</a></li>
<li class="toctree-l3"><a class="reference internal" href="#softmask-track-1">softmask_track_1</a></li>
<li class="toctree-l3"><a class="reference internal" href="#softmask-track-2">softmask_track_2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sra2fastq-pe">sra2fastq_PE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sra2fastq-se">sra2fastq_SE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#star-align">star_align</a></li>
<li class="toctree-l3"><a class="reference internal" href="#star-index">star_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#star-quant">star_quant</a></li>
<li class="toctree-l3"><a class="reference internal" href="#trackhub">trackhub</a></li>
<li class="toctree-l3"><a class="reference internal" href="#trackhub-index">trackhub_index</a></li>
<li class="toctree-l3"><a class="reference internal" href="#trim-galore-pe">trim_galore_PE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#trim-galore-se">trim_galore_SE</a></li>
<li class="toctree-l3"><a class="reference internal" href="#twobit">twobit</a></li>
<li class="toctree-l3"><a class="reference internal" href="#txi-count-matrix">txi_count_matrix</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="schemas.html">All configurable options</a></li>
<li class="toctree-l2"><a class="reference internal" href="CHANGELOG.html">Changelog</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="extensive_docs.html">Extensive docs</a> &raquo;</li>
        
      <li>Per rule explanation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/all_rules.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="per-rule-explanation">
<h1>Per rule explanation<a class="headerlink" href="#per-rule-explanation" title="Permalink to this headline">¶</a></h1>
<p>This is an automatically generated list of all supported rules, their docstrings, and command. At the start of each workflow run a list is printed of which rules will be run. And while the workflow is running it prints which rules are being started and finished. This page is here to give an explanation to the user about what each rule does, and for developers to find what is, and isn’t yet supported.</p>
<div class="section" id="bam2cram">
<h2>bam2cram<a class="headerlink" href="#bam2cram" title="Permalink to this headline">¶</a></h2>
<p>Convert a bam file to the more compressed cram format.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools view -T {input.assembly} -C {input.bam} -@ {threads} &gt; {output} 2&gt; {log}
</pre></div>
</div>
</div>
<div class="section" id="bam-bigwig">
<h2>bam_bigwig<a class="headerlink" href="#bam-bigwig" title="Permalink to this headline">¶</a></h2>
<p>Convert a bam file into a bigwig file</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bamCoverage --bam {input.bam} --outFileName {output} --numberOfProcessors {threads} \
{params} --verbose &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="bam-stranded-bigwig">
<h2>bam_stranded_bigwig<a class="headerlink" href="#bam-stranded-bigwig" title="Permalink to this headline">¶</a></h2>
<p>Convert a bam file into two bigwig files, one for each strand.
TODO: split this rule into two or better: forward &amp; backward logic</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>direction1=forward
direction2=reverse
if [ {params.strandedness} == &#39;reverse&#39; ]; then
    direction1=reverse
    direction2=forward
fi

bamCoverage --bam {input.bam} --outFileName {output.forwards} --filterRNAstrand $direction1 \
--numberOfProcessors {threads} {params.flags} --verbose &gt;&gt; {log} 2&gt;&amp;1 &amp;&amp;
bamCoverage --bam {input.bam} --outFileName {output.reverses} --filterRNAstrand $direction2 \
--numberOfProcessors {threads} {params.flags} --verbose &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="bedgraph-bigwig">
<h2>bedgraph_bigwig<a class="headerlink" href="#bedgraph-bigwig" title="Permalink to this headline">¶</a></h2>
<p>Convert a bedgraph file into a bigwig.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>awk -v OFS=&#39;\\t&#39; &#39;{{print $1, $2, $3, $4}}&#39; {input.bedgraph} | sed &#39;/experimental/d&#39; |
bedSort /dev/stdin {output.tmp} &gt; {log} 2&gt;&amp;1
bedGraphToBigWig {output.tmp} {input.genome_size} {output.out} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="bedgraphish-to-bedgraph">
<h2>bedgraphish_to_bedgraph<a class="headerlink" href="#bedgraphish-to-bedgraph" title="Permalink to this headline">¶</a></h2>
<p>Convert the bedgraph-ish file generated by genrich into a true bedgraph file.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>splits=$(grep -Pno &quot;([^\/]*)(?=\.bam)&quot; {input})
splits=($splits)
lncnt=$(wc -l {input} | echo &quot;$(grep -Po &quot;.*(?=\ )&quot;)&quot;:)
splits+=($lncnt)

counter=1
for split in &quot;${{splits[@]::${{#splits[@]}}-1}}&quot;;
do
    filename=$(grep -Po &quot;(?&lt;=:).*&quot; &lt;&lt;&lt; $split);
    if [[ $filename =~ {wildcards.sample} ]]; then
        startnr=$(grep -Po &quot;.*(?=\:)&quot; &lt;&lt;&lt; $split);
        endnr=$(grep -Po &quot;.*(?=\:)&quot; &lt;&lt;&lt; ${{splits[counter]}});

        lines=&quot;NR&gt;=startnr &amp;&amp; NR&lt;=endnr {{ print \$1, \$2, \$3, \$4 }}&quot;
        lines=${{lines/startnr/$((startnr + 2))}}
        lines=${{lines/endnr/$((endnr - 1))}}

        awk &quot;$lines&quot; {input} &gt; {output}
    fi
    ((counter++))
done
</pre></div>
</div>
</div>
<div class="section" id="blind-clustering">
<h2>blind_clustering<a class="headerlink" href="#blind-clustering" title="Permalink to this headline">¶</a></h2>
<p>Create a sample distance matrix plot per assembly</p>
</div>
<div class="section" id="bowtie2-align">
<h2>bowtie2_align<a class="headerlink" href="#bowtie2-align" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a genome (index) with bowtie2, and pipe the output to the required sorter(s).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bowtie2 {params.params} --threads {threads} -x {input.index}{wildcards.assembly} {params.input} 2&gt; {log} | tee {output} 1&gt; /dev/null 2&gt;&gt; {log}
</pre></div>
</div>
</div>
<div class="section" id="bowtie2-index">
<h2>bowtie2_index<a class="headerlink" href="#bowtie2-index" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for bowtie2. This index is required for alignment.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bowtie2-build {params} --threads {threads} {input} {output}/{wildcards.assembly} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="bwa-index">
<h2>bwa_index<a class="headerlink" href="#bwa-index" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for bwa (mem). This index is required for alignment.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bwa index -p {params.prefix} {params.params} {input} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="bwa-mem">
<h2>bwa_mem<a class="headerlink" href="#bwa-mem" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a genome (index) with bwa-mem, and pipe the output to the required sorter(s).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bwa mem {params.params} -t {threads} {params.index_dir} {input.reads} 2&gt; {log} | tee {output} 1&gt; /dev/null 2&gt;&gt; {log}
</pre></div>
</div>
</div>
<div class="section" id="call-peak-genrich">
<h2>call_peak_genrich<a class="headerlink" href="#call-peak-genrich" title="Permalink to this headline">¶</a></h2>
<p>Call peaks with genrich based on the pileup.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>Genrich -P -f {input.log} -o {output.narrowpeak} {params} -v &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="complement-blacklist">
<h2>complement_blacklist<a class="headerlink" href="#complement-blacklist" title="Permalink to this headline">¶</a></h2>
<p>Take the complement of the blacklist. We need this complement to tell samtools
to only keep reads that are in the complement of the blacklist.</p>
</div>
<div class="section" id="computematrix">
<h2>computeMatrix<a class="headerlink" href="#computematrix" title="Permalink to this headline">¶</a></h2>
<p>Pre-compute correlations between bams using deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>computeMatrix scale-regions -S {input.bw} {params.labels} -R {params.annotation} \
-p {threads} -b 2000 -a 500 -o {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="counts-matrix">
<h2>counts_matrix<a class="headerlink" href="#counts-matrix" title="Permalink to this headline">¶</a></h2>
<p>Merge gene counts per assembly</p>
<p>From <a class="reference external" href="https://github.com/snakemake-workflows/rna-seq-star-deseq2">https://github.com/snakemake-workflows/rna-seq-star-deseq2</a></p>
</div>
<div class="section" id="create-snap-object">
<h2>create_SNAP_object<a class="headerlink" href="#create-snap-object" title="Permalink to this headline">¶</a></h2>
<p>Create a snapobject for each BAM file.</p>
<p>These snapobjects can be merged later using snaptools in R.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>snaptools snap-pre --input-file={input.bams} --output-snap={output} \
--genome-name={wildcards.assembly} --genome-size={input.genome_size} \
{params.params} {params.chrm} {params.mapq} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="create-bins-snap-object">
<h2>create_bins_SNAP_object<a class="headerlink" href="#create-bins-snap-object" title="Permalink to this headline">¶</a></h2>
<p>Add a Binned genome matrix to the SNAPobject, after which it is renamed and moved
to the Snapfiles folder for downstream analysis in R using Snaptools</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>snaptools snap-add-bmat --snap-file={input} {params} &gt; {log} 2&gt;&amp;1
echo &quot;bmat added, moving file&quot; &gt;&gt; {log} 2&gt;&amp;1
mv {input} {output} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="cytoband">
<h2>cytoband<a class="headerlink" href="#cytoband" title="Permalink to this headline">¶</a></h2>
<p>Generate a cytoband track for each assembly</p>
<p>source: <a class="reference external" href="http://genomewiki.ucsc.edu/index.php/Assembly_Hubs#Cytoband_Track">http://genomewiki.ucsc.edu/index.php/Assembly_Hubs#Cytoband_Track</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>cat {input.sizes} |
bedSort /dev/stdin /dev/stdout |
awk &#39;{{print $1,0,$2,$1,&quot;gneg&quot;}}&#39; &gt; {output.cytoband_bd}

bedToBigBed -type=bed4 {output.cytoband_bd} -as={params.schema} \
{input.sizes} {output.cytoband_bb} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="decoy-transcripts">
<h2>decoy_transcripts<a class="headerlink" href="#decoy-transcripts" title="Permalink to this headline">¶</a></h2>
<p>Generate decoy_transcripts.txt for Salmon indexing</p>
<p>script source: <a class="reference external" href="https://github.com/COMBINE-lab/SalmonTools">https://github.com/COMBINE-lab/SalmonTools</a></p>
</div>
<div class="section" id="deseq2">
<h2>deseq2<a class="headerlink" href="#deseq2" title="Permalink to this headline">¶</a></h2>
<p>Differential gene expression analysis with DESeq2.</p>
</div>
<div class="section" id="fastqc">
<h2>fastqc<a class="headerlink" href="#fastqc" title="Permalink to this headline">¶</a></h2>
<p>Generate quality control report for fastq files.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>fastqc {input} -O {params} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="featurecounts">
<h2>featureCounts<a class="headerlink" href="#featurecounts" title="Permalink to this headline">¶</a></h2>
<p>Use featureCounts to generate the fraction reads in peaks score
(frips/assigned reads). See: <a class="reference external" href="https://www.biostars.org/p/337872/">https://www.biostars.org/p/337872/</a> and
<a class="reference external" href="https://www.biostars.org/p/228636/">https://www.biostars.org/p/228636/</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span># Make a custom &quot;SAF&quot; file which featureCounts needs:
awk &#39;BEGIN{{FS=OFS=&quot;\t&quot;; print &quot;GeneID\tChr\tStart\tEnd\tStrand&quot;}}{{print $4, $1, $2+1, $3, &quot;.&quot;}}&#39; \
{input.peak} 1&gt; {output.tmp_saf} 2&gt; {log}

# run featureCounts
featureCounts -T {threads} -p -a {output.tmp_saf} -F SAF -o {output.real_out} {input.bam} &gt; {log} 2&gt;&amp;1

# move the summary to qc directory
mv $(dirname {output.real_out})/$(basename {output.summary}) {output.summary}
</pre></div>
</div>
</div>
<div class="section" id="gcpercent">
<h2>gcPercent<a class="headerlink" href="#gcpercent" title="Permalink to this headline">¶</a></h2>
<p>Generate a gc content track</p>
<p>source: <a class="reference external" href="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/gc5Base/">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/gc5Base/</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>        hgGcPercent -wigOut -doGaps -file=stdout -win=5 -verbose=0 {wildcards.assembly} {input.twobit} \
| gzip &gt; {output.gcpcvar}

        wigToBigWig {output.gcpcvar} {input.sizes} {output.gcpc} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="genrich-pileup">
<h2>genrich_pileup<a class="headerlink" href="#genrich-pileup" title="Permalink to this headline">¶</a></h2>
<p>Generate the pileup. We do this separately from peak-calling since these two
processes have a very different computational footprint.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>input=$(echo {input.reps} | tr &#39; &#39; &#39;,&#39;)
Genrich -X -t $input -f {output.log} {params.control} -k {output.bedgraphish} \
{params.params} -v &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="get-genome">
<h2>get_genome<a class="headerlink" href="#get-genome" title="Permalink to this headline">¶</a></h2>
<p>Download a genome through genomepy.
Additionally downloads the gene annotation if required downstream.</p>
<p>If assemblies with the same name can be downloaded from multiple providers,
a provider may be specified in the config (example: provider: NCBI). Otherwise,
each provider will be tried in turn, stopping at the first success.</p>
<p>Automatically turns on/off plugins.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span># turn off plugins and reset on exit. delete temp files on exit.
active_plugins=$(genomepy config show | grep -Po &#39;(?&lt;=- ).*&#39; | paste -s -d, -) || echo &quot;&quot;
trap &quot;genomepy plugin enable {{$active_plugins,}} &gt;&gt; {log} 2&gt;&amp;1; rm -f {params.temp}&quot; EXIT
genomepy plugin disable {{blacklist,bowtie2,bwa,star,gmap,hisat2,minimap2}} &gt;&gt; {log} 2&gt;&amp;1
genomepy plugin enable {{blacklist,}} &gt;&gt; {log} 2&gt;&amp;1

# download the genome and attempt to download the annotation and blacklist
if [[ ! {params.provider} = None  ]]; then
    genomepy install --genomes_dir {params.dir} {wildcards.assembly} {params.provider} --annotation &gt;&gt; {log} 2&gt;&amp;1
else
    genomepy install --genomes_dir {params.dir} {wildcards.assembly} Ensembl --annotation &gt;&gt; {log} 2&gt;&amp;1 ||
    genomepy install --genomes_dir {params.dir} {wildcards.assembly} UCSC    --annotation &gt;&gt; {log} 2&gt;&amp;1 ||
    genomepy install --genomes_dir {params.dir} {wildcards.assembly} NCBI    --annotation &gt;&gt; {log} 2&gt;&amp;1
fi

# unzip annotation if downloaded and gzipped
if [ -f {params.gtf}.gz ]; then
    gunzip -f {params.gtf}.gz &gt;&gt; {log} 2&gt;&amp;1
fi

# if assembly has no annotation, or annotation has no genes, throw an warning
if [ ! -f {params.gtf} ] || $(grep -q &quot;No genes found&quot; {log}); then
    echo &#39;\nEmpty or no annotation found for {wildcards.assembly}.\n&#39; | tee -a {log}

    # if an annotation is required, make it an error and exit.
    if $(echo {output} | grep -q annotation.gtf); then
        echo&#39;\nSelect a different assembly or provide an annotation file manually.\n\n&#39; | tee -a {log}
        exit 1
    fi
fi
</pre></div>
</div>
</div>
<div class="section" id="get-transcripts">
<h2>get_transcripts<a class="headerlink" href="#get-transcripts" title="Permalink to this headline">¶</a></h2>
<p>Generate transcripts.fasta using gffread.</p>
<p>Requires genome.fa and annotation.gtf (with matching chromosome/scaffold names)</p>
</div>
<div class="section" id="hisat2-align">
<h2>hisat2_align<a class="headerlink" href="#hisat2-align" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a genome (index) with hisat2, and pipe the output to the required sorter(s).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>hisat2 {params.params} --threads {threads} -x {input.index}{wildcards.assembly} {params.input} 2&gt; {log} | tee {output} 1&gt; /dev/null 2&gt;&gt; {log}
</pre></div>
</div>
</div>
<div class="section" id="hisat2-index">
<h2>hisat2_index<a class="headerlink" href="#hisat2-index" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for hisat2. This index is required for alignment.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>hisat2-build {params} -p {threads} {input} {output}/{wildcards.assembly} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="hmmratac">
<h2>hmmratac<a class="headerlink" href="#hmmratac" title="Permalink to this headline">¶</a></h2>
<p>Call gappedpeaks with HMMRATAC.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>HMMRATAC --bedgraph true -o {params.basename} {params.hmmratac_params} -Xmx22G -b {input.bam} -i {input.bam_index} -g {input.genome_size} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="hmmratac-genome-info">
<h2>hmmratac_genome_info<a class="headerlink" href="#hmmratac-genome-info" title="Permalink to this headline">¶</a></h2>
<p>Generate the ‘genome info’ that hmmratac requires for peak calling.
<a class="reference external" href="https://github.com/LiuLabUB/HMMRATAC/issues/17">https://github.com/LiuLabUB/HMMRATAC/issues/17</a></p>
<p>TODO: isnt this just .fa.sizes?</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools view -H {input} 2&gt;  {log} | grep SQ 2&gt;&gt; {log} | cut -f 2-3 2&gt;&gt; {log} | cut -d &#39;:&#39; -f 2   2&gt;&gt; {log} | cut -f 1        &gt; {output.tmp1} 2&gt; {log}
samtools view -H {input} 2&gt;&gt; {log} | grep SQ 2&gt;&gt; {log} | cut -f 2-3 2&gt;&gt; {log} | cut -d &#39;:&#39; -f 2,3 2&gt;&gt; {log} | cut -d &#39;:&#39; -f 2 &gt; {output.tmp2} 2&gt; {log}
paste {output.tmp1} {output.tmp2} &gt; {output.out} 2&gt; {log}
</pre></div>
</div>
</div>
<div class="section" id="id2sra">
<h2>id2sra<a class="headerlink" href="#id2sra" title="Permalink to this headline">¶</a></h2>
<p>Download the SRA of a sample by its unique identifier.</p>
<p>Tries first downloading with the faster ascp protocol, if that fails it
falls back on the slower http protocol.</p>
</div>
<div class="section" id="idr">
<h2>idr<a class="headerlink" href="#idr" title="Permalink to this headline">¶</a></h2>
<p>Combine replicates based on the irreproducible discovery rate (IDR). Can only handle two replicates.
For more than two replicates use fisher’s method.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>                if [ &quot;{params.nr_reps}&quot; == &quot;1&quot; ]; then
cp {input} {output}
                else
idr --samples {input} {params.rank} --output-file {output} &gt; {log} 2&gt;&amp;1
                fi
</pre></div>
</div>
</div>
<div class="section" id="insert-size-metrics">
<h2>insert_size_metrics<a class="headerlink" href="#insert-size-metrics" title="Permalink to this headline">¶</a></h2>
<p>Get insert size metrics from a (paired-end) bam. This score is then used by
MultiQC in the report.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>picard CollectInsertSizeMetrics INPUT={input} \
OUTPUT={output.tsv} H={output.pdf} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="keep-mates">
<h2>keep_mates<a class="headerlink" href="#keep-mates" title="Permalink to this headline">¶</a></h2>
<p>In-house script that, after alignment, removes the information that reads are paired.
This can be beneficial when peak calling with macs2 when shifting + extending, since
macs2 in this case only keeps the first in pair.</p>
</div>
<div class="section" id="linked-txome">
<h2>linked_txome<a class="headerlink" href="#linked-txome" title="Permalink to this headline">¶</a></h2>
<p>Generate a linked transcriptome for tximeta</p>
<p>Also creates a symlink to the gtf in an Ensembl format (required by tximeta)</p>
<p>Required to converting salmon output (estimated transcript abundances) to gene counts</p>
</div>
<div class="section" id="macs2-callpeak">
<h2>macs2_callpeak<a class="headerlink" href="#macs2-callpeak" title="Permalink to this headline">¶</a></h2>
<p>Call peaks using macs2.
Macs2 requires a genome size, which we estimate from the amount of unique kmers of the average read length.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span># extract the kmer size, and get the effective genome size from it
kmer_size=$(unzip -p {{input.fastqc}} {{params.name}}_trimmed_fastqc/fastqc_data.txt  | grep -P -o &#39;(?&lt;=Sequence length\\t).*&#39; | grep -P -o &#39;\d+$&#39;);

echo &quot;preparing to run unique-kmers.py with -k $kmer_size&quot; &gt;&gt; {{log}}
GENSIZE=$(unique-kmers.py {{params.genome}} -k $kmer_size --quiet 2&gt;&amp;1 | grep -P -o &#39;(?&lt;=\.fa: ).*&#39;);
echo &quot;kmer size: $kmer_size, and effective genome size: $GENSIZE&quot; &gt;&gt; {{log}}

# call peaks
macs2 callpeak --bdg -t {{input.bam}} {{params.control}} --outdir {config[&#39;result_dir&#39;]}/macs2/ -n {{wildcards.assembly}}-{{wildcards.sample}} \
{{params.macs_params}} -g $GENSIZE -f {{params.format}} &gt;&gt; {{log}} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="macs-bdgcmp">
<h2>macs_bdgcmp<a class="headerlink" href="#macs-bdgcmp" title="Permalink to this headline">¶</a></h2>
<p>Prepare p-value files for rule macs_cmbreps</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>macs2 bdgcmp -t {input.treatment} -c {input.control} -m qpois -o {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="macs-cmbreps">
<h2>macs_cmbreps<a class="headerlink" href="#macs-cmbreps" title="Permalink to this headline">¶</a></h2>
<p>Combine replicates through Fisher’s method</p>
<p>(Link original peakfile in replicate_processed if there is only 1 sample for a condition)</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>                    if [ &quot;{params.nr_reps}&quot; == &quot;1&quot; ]; then
touch {output.tmpbdg} {output.tmppeaks}
mkdir -p $(dirname {output.peaks}); cp {input.treatment} {output.peaks}
                    else
macs2 cmbreps -i {input.bdgcmp} -o {output.tmpbdg} -m fisher &gt; {log} 2&gt;&amp;1
macs2 {params.function} {params.config} -i {output.tmpbdg} -o {output.tmppeaks} &gt;&gt; {log} 2&gt;&amp;1
cat {output.tmppeaks} | tail -n +2 &gt; {output.peaks}
                    fi
</pre></div>
</div>
</div>
<div class="section" id="mark-duplicates">
<h2>mark_duplicates<a class="headerlink" href="#mark-duplicates" title="Permalink to this headline">¶</a></h2>
<p>Mark all duplicate reads in a bam file with picard MarkDuplicates</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>picard MarkDuplicates {params} INPUT={input} \
OUTPUT={output.bam} METRICS_FILE={output.metrics} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="merge-replicates">
<h2>merge_replicates<a class="headerlink" href="#merge-replicates" title="Permalink to this headline">¶</a></h2>
<p>Merge replicates (fastqs) simply by concatenating the files. We also change the name of the read headers to
contain the name of the original replicate.</p>
<p>Must happen after trimming due to trim-galore’s automatic adapter trimming method</p>
<p>If a replicate has only 1 sample in it, simply move the file.</p>
</div>
<div class="section" id="mt-nuc-ratio-calculator">
<h2>mt_nuc_ratio_calculator<a class="headerlink" href="#mt-nuc-ratio-calculator" title="Permalink to this headline">¶</a></h2>
<p>Estimate the amount of nuclear and mitochondrial reads in a sample. This metric
is especially important in ATAC-seq experiments where mitochondrial DNA can
be overrepresented. Reads are classified as mitochondrial reads if they map
against either “chrM” or “MT”.</p>
<p>These values are aggregated and displayed in the MultiQC report.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>mtnucratio {input.bam} {params.mitochondria}
</pre></div>
</div>
</div>
<div class="section" id="multibamsummary">
<h2>multiBamSummary<a class="headerlink" href="#multibamsummary" title="Permalink to this headline">¶</a></h2>
<p>Pre-compute a bam summary with deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>multiBamSummary bins --bamfiles {input.bams} -out {output} {params} -p {threads} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="multiqc">
<h2>multiqc<a class="headerlink" href="#multiqc" title="Permalink to this headline">¶</a></h2>
<p>Aggregate all the quality control metrics for every sample into a single multiqc report.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>multiqc {input.files} -o {params.dir} -n multiqc_{wildcards.assembly}.html \
--config {input.schema}                                                    \
--config {input.header}                                                    \
--sample-names {input.sample_names}                                        \
--sample-filters {input.filter_buttons}                                    \
--cl_config &quot;extra_fn_clean_exts: [                                        \
    {{&#39;pattern&#39;: ^.*{wildcards.assembly}-, &#39;type&#39;: &#39;regex&#39;}},              \
    {{&#39;pattern&#39;: {params.fqext1},          &#39;type&#39;: &#39;regex&#39;}},              \
    ]&quot; &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="multiqc-filter-buttons">
<h2>multiqc_filter_buttons<a class="headerlink" href="#multiqc-filter-buttons" title="Permalink to this headline">¶</a></h2>
<p>Generate filter buttons.</p>
</div>
<div class="section" id="multiqc-header-info">
<h2>multiqc_header_info<a class="headerlink" href="#multiqc-header-info" title="Permalink to this headline">¶</a></h2>
<p>Generate a multiqc header file with contact info and date of multiqc generation.</p>
</div>
<div class="section" id="multiqc-rename-buttons">
<h2>multiqc_rename_buttons<a class="headerlink" href="#multiqc-rename-buttons" title="Permalink to this headline">¶</a></h2>
<p>Generate rename buttons for the multiqc report.</p>
</div>
<div class="section" id="multiqc-samplesconfig">
<h2>multiqc_samplesconfig<a class="headerlink" href="#multiqc-samplesconfig" title="Permalink to this headline">¶</a></h2>
<p>Add a section in the multiqc report that reports the samples.tsv and config.yaml</p>
</div>
<div class="section" id="multiqc-schema">
<h2>multiqc_schema<a class="headerlink" href="#multiqc-schema" title="Permalink to this headline">¶</a></h2>
<p>Generate a multiqc config schema. Used for the ordering of modules.</p>
</div>
<div class="section" id="peak-bigpeak">
<h2>peak_bigpeak<a class="headerlink" href="#peak-bigpeak" title="Permalink to this headline">¶</a></h2>
<p>Convert a narrowpeak file into a bignarrowpeak file.
<a class="reference external" href="https://genome-source.gi.ucsc.edu/gitlist/kent.git/tree/master/src/hg/lib/">https://genome-source.gi.ucsc.edu/gitlist/kent.git/tree/master/src/hg/lib/</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span># keep first 10 columns, idr adds extra columns we do not need for our bigpeak
cut -d$&#39;\t&#39; -f 1-{params.columns} {input.narrowpeak} |
awk -v OFS=&quot;\t&quot; &#39;{{$5=$5&gt;1000?1000:$5}} {{print}}&#39; |
bedSort /dev/stdin {output.tmp} &gt; {log} 2&gt;&amp;1

bedToBigBed -type={params.type} -as={params.schema} {output.tmp} \
{input.genome_size} {output.out} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="peak-union">
<h2>peak_union<a class="headerlink" href="#peak-union" title="Permalink to this headline">¶</a></h2>
<p>TODO improve + explain</p>
</div>
<div class="section" id="plotcorrelation">
<h2>plotCorrelation<a class="headerlink" href="#plotcorrelation" title="Permalink to this headline">¶</a></h2>
<p>Calculate the correlation between bams with deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>plotCorrelation --corData {input} --outFileCorMatrix {output} -c spearman -p heatmap &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="plotfingerprint">
<h2>plotFingerprint<a class="headerlink" href="#plotfingerprint" title="Permalink to this headline">¶</a></h2>
<p>Plot the “fingerprint” of your bams, using deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>plotFingerprint -b {input.bams} {params} --outRawCounts {output} -p {threads} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="plotpca">
<h2>plotPCA<a class="headerlink" href="#plotpca" title="Permalink to this headline">¶</a></h2>
<p>Plot a PCA between bams using deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>plotPCA --corData {input} --outFileNameData {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="plotprofile">
<h2>plotProfile<a class="headerlink" href="#plotprofile" title="Permalink to this headline">¶</a></h2>
<p>Plot the so-called profile using deeptools.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>plotProfile -m {input} --outFileName {output.img} --outFileNameData {output.file} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="renamefastq-pe">
<h2>renamefastq_PE<a class="headerlink" href="#renamefastq-pe" title="Permalink to this headline">¶</a></h2>
<p>Create symlinks to fastqs with incorrect fqexts (default R1/R2).
Forward and reverse samples will be switched if forward/reverse names are not
lexicographically ordered.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>ln {input[0]} {output[0]}
ln {input[1]} {output[1]}
</pre></div>
</div>
</div>
<div class="section" id="salmon-decoy-aware-index">
<h2>salmon_decoy_aware_index<a class="headerlink" href="#salmon-decoy-aware-index" title="Permalink to this headline">¶</a></h2>
<p>Make a decoy aware transcript index for Salmon.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>salmon index -t {input.transcripts} --decoys {input.decoy_transcripts} -i {output} {params} \
--threads {threads} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="salmon-index">
<h2>salmon_index<a class="headerlink" href="#salmon-index" title="Permalink to this headline">¶</a></h2>
<p>Make a transcriptomic index for Salmon.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>salmon index -t {input} -i {output} {params} --threads {threads} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="salmon-quant">
<h2>salmon_quant<a class="headerlink" href="#salmon-quant" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a transcriptome (index) with Salmon (mapping-based mode) and output a quantification file per sample.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>salmon quant -i {input.index} -l A {params.input} {params.params} -o {output.dir} \
--threads {threads} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="sambamba-sort">
<h2>sambamba_sort<a class="headerlink" href="#sambamba-sort" title="Permalink to this headline">¶</a></h2>
<p>Sort the result of alignment or sieving with the sambamba sorter.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sambamba view --nthreads {threads} -f bam  {input} -o /dev/stdout  2&gt; {log} |
sambamba sort --nthreads {threads} {params} /dev/stdin -o {output}  2&gt; {log}
</pre></div>
</div>
</div>
<div class="section" id="samtools-index">
<h2>samtools_index<a class="headerlink" href="#samtools-index" title="Permalink to this headline">¶</a></h2>
<p>Create an index of a bam file which can be used for e.g. visualization.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools index {params} {input} {output}
</pre></div>
</div>
</div>
<div class="section" id="samtools-index-cram">
<h2>samtools_index_cram<a class="headerlink" href="#samtools-index-cram" title="Permalink to this headline">¶</a></h2>
<p>Generate the index for a cram file.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>samtools index {input} {output} &gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="samtools-presort">
<h2>samtools_presort<a class="headerlink" href="#samtools-presort" title="Permalink to this headline">¶</a></h2>
<p>(Pre)sort the result of alignment with the samtools sorter.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span># we set this trap to remove temp files when prematurely ending the rule
trap &quot;rm -f {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp*bam&quot; INT;

samtools sort {params.sort_order} -@ {params.threads} {input} -o {output} \
-T {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp 2&gt; {log}
</pre></div>
</div>
</div>
<div class="section" id="samtools-sort">
<h2>samtools_sort<a class="headerlink" href="#samtools-sort" title="Permalink to this headline">¶</a></h2>
<p>Sort the result of sieving with the samtools sorter.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span># we set this trap to remove temp files when prematurely ending the rule
trap &quot;rm -f {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp*bam&quot; INT;

samtools sort {params.sort_order} -@ {params.threads} {input} -o {output} \
-T {params.out_dir}/{wildcards.assembly}-{wildcards.sample}.tmp 2&gt; {log}
</pre></div>
</div>
</div>
<div class="section" id="samtools-stats">
<h2>samtools_stats<a class="headerlink" href="#samtools-stats" title="Permalink to this headline">¶</a></h2>
<p>Get general stats from bam files like percentage mapped.</p>
</div>
<div class="section" id="setup-blacklist">
<h2>setup_blacklist<a class="headerlink" href="#setup-blacklist" title="Permalink to this headline">¶</a></h2>
<p>Combine the encode blacklist with mitochrondrial dna depending on config.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sortBed -faidx {input.sizes} -i {input.blacklist} |
complementBed -i stdin -g {input.sizes} &gt; {output} 2&gt; {log}
</pre></div>
</div>
</div>
<div class="section" id="sieve-bam">
<h2>sieve_bam<a class="headerlink" href="#sieve-bam" title="Permalink to this headline">¶</a></h2>
<p>“Sieve” a bam.</p>
<p>This rule does (at least) one of these:</p>
<ul>
<li><p>filtering on minimum mapping quality</p></li>
<li><p>tn5 shift adjustment</p></li>
<li><p>remove multimappers</p></li>
<li><p>remove reads inside the blacklist
.. code-block:: guess</p>
<blockquote>
<div><p>samtools view -h {params.prim_align} {params.minqual} {params.blacklist} {input.bam} {params.atacshift} |
samtools view -b &gt; {output} 2&gt; {log}</p>
</div></blockquote>
</li>
</ul>
</div>
<div class="section" id="softmask-track-1">
<h2>softmask_track_1<a class="headerlink" href="#softmask-track-1" title="Permalink to this headline">¶</a></h2>
<p>Generate a track of all softmasked regions</p>
<p>source: <a class="reference external" href="https://github.com/Gaius-Augustus/MakeHub/blob/master/make_hub.py">https://github.com/Gaius-Augustus/MakeHub/blob/master/make_hub.py</a></p>
</div>
<div class="section" id="softmask-track-2">
<h2>softmask_track_2<a class="headerlink" href="#softmask-track-2" title="Permalink to this headline">¶</a></h2>
<p>Generate a track of all softmasked regions</p>
<p>source: <a class="reference external" href="https://github.com/Gaius-Augustus/MakeHub/blob/master/make_hub.py">https://github.com/Gaius-Augustus/MakeHub/blob/master/make_hub.py</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>bedSort {input.mask_unsorted} {output.maskbed} &gt;&gt; {log} 2&gt;&amp;1

bedToBigBed -type=bed3 {output.maskbed} {input.sizes} {output.mask} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="sra2fastq-pe">
<h2>sra2fastq_PE<a class="headerlink" href="#sra2fastq-pe" title="Permalink to this headline">¶</a></h2>
<p>Downloaded (raw) SRAs are converted to paired-end fastq files.
Forward and reverse samples will be switched if forward/reverse names are not lexicographically ordered.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span># setup tmp dir
tmpdir={config[&#39;sra_dir&#39;]}/tmp/{{wildcards.sample}}
mkdir -p $tmpdir; trap &quot;rm -rf $tmpdir&quot; EXIT

# dump to tmp dir
parallel-fastq-dump -s {{input}}/* -O $tmpdir {config[&#39;split&#39;]} \
--threads {{threads}} --gzip &gt;&gt; {{log}} 2&gt;&amp;1

# rename files and move to output dir
for f in $(ls -1q $tmpdir | grep -oP &quot;^[^_]+&quot; | uniq); do
    dst_1={config[&#39;fastq_dir&#39;]}/{{wildcards.sample}}_{config[&#39;fqext1&#39;]}.{config[&#39;fqsuffix&#39;]}.gz
    dst_2={config[&#39;fastq_dir&#39;]}/{{wildcards.sample}}_{config[&#39;fqext2&#39;]}.{config[&#39;fqsuffix&#39;]}.gz
    cat &quot;${{{{tmpdir}}}}/${{{{f}}}}_pass_1.fastq.gz&quot; &gt;&gt; $dst_1
    cat &quot;${{{{tmpdir}}}}/${{{{f}}}}_pass_2.fastq.gz&quot; &gt;&gt; $dst_2
done
</pre></div>
</div>
</div>
<div class="section" id="sra2fastq-se">
<h2>sra2fastq_SE<a class="headerlink" href="#sra2fastq-se" title="Permalink to this headline">¶</a></h2>
<p>Downloaded (raw) SRAs are converted to single-end fastq files.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span># setup tmp dir
tmpdir={config[&#39;sra_dir&#39;]}/tmp/{{wildcards.sample}}
mkdir -p $tmpdir; trap &quot;rm -rf $tmpdir&quot; EXIT

# dump to tmp dir
parallel-fastq-dump -s {{input}}/* -O $tmpdir {config[&#39;splot&#39;]} \
--threads {{threads}} --gzip &gt;&gt; {{log}} 2&gt;&amp;1

# rename file and move to output dir
for f in $(ls -1q $tmpdir | grep -oP &quot;^[^_]+&quot; | uniq); do
    dst={config[&#39;fastq_dir&#39;]}/{{wildcards.sample}}.{config[&#39;fqsuffix&#39;]}.gz
    cat &quot;${{{{tmpdir}}}}/${{{{f}}}}_pass.fastq.gz&quot; &gt;&gt; $dst
done
</pre></div>
</div>
</div>
<div class="section" id="star-align">
<h2>star_align<a class="headerlink" href="#star-align" title="Permalink to this headline">¶</a></h2>
<p>Align reads against a genome (index) with STAR, and pipe the output to the required sorter(s).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>trap &quot;find {log} -type f ! -name Log* -exec rm {{}} \;&quot; EXIT
mkdir -p {log}
mkdir -p {output.dir}

STAR --genomeDir {input.index} --readFilesIn {params.input} --readFilesCommand gunzip -c \
--quantMode GeneCounts --outSAMtype BAM Unsorted --outStd BAM_Unsorted \
--outFileNamePrefix {log}/ --outTmpDir {output.dir}/STARtmp \
--runThreadN {threads} {params.params} &gt; {output.pipe} 2&gt; {log}/Log.stderr.out

# move all non-log files to output directory (this way the log files are kept on error)
find {log} -type f ! -name Log* -exec mv {{}} {output.dir} \;
</pre></div>
</div>
</div>
<div class="section" id="star-index">
<h2>star_index<a class="headerlink" href="#star-index" title="Permalink to this headline">¶</a></h2>
<p>Make a genome index for STAR.</p>
<p>Troubleshooting:
1) sufficient disk space?
2) increase the RAM available (–limitGenomeGenerateRAM)
3) reduce the number of threads (snakemake -j 5)
4) reduce accuracy (–genomeSAsparseD 2)</p>
<p>For example, in your config.yaml, set aligner/quantifier:
aligner:
star:
index: –limitGenomeGenerateRAM 60000000000 –genomeSAsparseD 1</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>            function log2 {{
local x=0
for (( y=$1-1 ; $y &gt; 0; y &gt;&gt;= 1 )) ; do
    let x=$x+1
done
echo $x
            }}

            # set genome dependent variables
            NBits=&quot;&quot;
            NBases=&quot;&quot;
            GenomeLength=$(awk -F&quot;\t&quot; &#39;{{x+=$2}}END{{printf &quot;%i&quot;, x}}&#39; {input.sizefile})
            NumberOfReferences=$(awk &#39;END{{print NR}}&#39; {input.sizefile})
            if [ $NumberOfReferences -gt 5000 ]; then
                # for large genomes, --genomeChrBinNbits should be scaled to min(18,log2[max(GenomeLength/NumberOfReferences,ReadLength)])
                # ReadLength is skipped here, as it is unknown
                LpR=$(log2 $((GenomeLength / NumberOfReferences)))
                NBits=&quot;--genomeChrBinNbits $(($LpR&lt;18 ? $LpR : 18))&quot;
                printf &quot;NBits: $NBits\n\n&quot; &gt;&gt; {log} 2&gt;&amp;1
            fi

            if [ $GenomeLength -lt 268435456 ]; then
                # for small genomes, --genomeSAindexNbases must be scaled down to min(14, log2(GenomeLength)/2-1)
                logG=$(( $(log2 $GenomeLength) / 2 - 1 ))
                NBases=&quot;--genomeSAindexNbases $(( $logG&lt;14 ? $logG : 14 ))&quot;
                printf &quot;NBases: $NBases\n\n&quot; &gt;&gt; {log} 2&gt;&amp;1
            fi

            mkdir -p {output}

            STAR --runMode genomeGenerate --genomeFastaFiles {input.genome} --sjdbGTFfile {input.gtf} \
            --genomeDir {output} --outFileNamePrefix {output}/ \
            --runThreadN {threads} $NBits $NBases {params} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="star-quant">
<h2>star_quant<a class="headerlink" href="#star-quant" title="Permalink to this headline">¶</a></h2>
<p>Quantify reads against a genome and transcriptome (index) with STAR and output a counts table per sample.</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>trap &quot;find {log} -type f ! -name Log* -exec rm {{}} \;&quot; EXIT
mkdir -p {log}
mkdir -p {output.dir}

STAR --genomeDir {input.index} --readFilesIn {params.input} --readFilesCommand gunzip -c \
--quantMode GeneCounts --outSAMtype None \
--outFileNamePrefix {log}/ --outTmpDir {output.dir}/STARtmp \
--runThreadN {threads} {params.params} &gt; {log}/Log.std_stderr.out 2&gt;&amp;1

# move all non-log files to output directory (this way the log files are kept on error)
find {log} -type f ! -name Log* -exec mv {{}} {output.dir} \;
</pre></div>
</div>
</div>
<div class="section" id="trackhub">
<h2>trackhub<a class="headerlink" href="#trackhub" title="Permalink to this headline">¶</a></h2>
<p>Generate a trackhub which has to be hosted on an web accessible location,
but can then be viewed through the UCSC genome browser.</p>
</div>
<div class="section" id="trackhub-index">
<h2>trackhub_index<a class="headerlink" href="#trackhub-index" title="Permalink to this headline">¶</a></h2>
<p>Generate a searchable annotation &amp; index for each assembly</p>
<p>source: <a class="reference external" href="https://genome.ucsc.edu/goldenPath/help/hubQuickStartSearch.html">https://genome.ucsc.edu/goldenPath/help/hubQuickStartSearch.html</a></p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span># generate annotation files
gtfToGenePred -geneNameAsName2 -genePredExt {input.gtf} {output.genePred} -infoOut={output.info} &gt;&gt; {log} 2&gt;&amp;1

genePredToBed {output.genePred} {output.genePredbed} &gt;&gt; {log} 2&gt;&amp;1

bedSort {output.genePredbed} {output.genePredbed} &gt;&gt; {log} 2&gt;&amp;1

bedToBigBed -extraIndex=name {output.genePredbed} {input.sizes} {output.genePredbigbed} &gt;&gt; {log} 2&gt;&amp;1

# generate searchable indexes (by 2: geneId, 8: proteinID, 9: geneName, 10: transcriptName and 1: transcriptID)
grep -v &quot;^#&quot; {output.info} | awk &#39;{{print $1, $2, $8, $9, $10, $1}}&#39; &gt; {output.indexinfo}

ixIxx {output.indexinfo} {output.ix} {output.ixx} &gt;&gt; {log} 2&gt;&amp;1
</pre></div>
</div>
</div>
<div class="section" id="trim-galore-pe">
<h2>trim_galore_PE<a class="headerlink" href="#trim-galore-pe" title="Permalink to this headline">¶</a></h2>
<p>Automated adapter detection, adapter trimming, and quality trimming through trim galore (paired-end).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>trim_galore --paired -j {threads} {params.config} -o $(dirname {output.r1}) {input.r1} {input.r2} &gt; {log} 2&gt;&amp;1

# now rename to proper output
for f in $(find &quot;$(dirname {output.r1})/&quot; -name &quot;{wildcards.sample}_*val_*.fq.gz&quot;); do
    mv &quot;$f&quot; &quot;$(echo &quot;$f&quot; | sed s/.fq/.{params.fqsuffix}/)&quot;; done
for f in $(find &quot;$(dirname {output.r1})/&quot; -maxdepth 1 -name &quot;{wildcards.sample}_*val_*.{params.fqsuffix}.gz&quot;); do
    mv &quot;$f&quot; &quot;$(echo &quot;$f&quot; | sed s/_val_./_trimmed/)&quot;; done

# move the trimming reports to qc directory
for f in $(find &quot;$(dirname {output.r1})/&quot; -name &quot;{wildcards.sample}_*.{params.fqsuffix}.gz_trimming_report.txt&quot;); do
    mv &quot;$f&quot; &quot;$(dirname {output.qc[0]})/$(basename $f)&quot;; done
</pre></div>
</div>
</div>
<div class="section" id="trim-galore-se">
<h2>trim_galore_SE<a class="headerlink" href="#trim-galore-se" title="Permalink to this headline">¶</a></h2>
<p>Automated adapter detection, adapter trimming, and quality trimming through trim galore (single-end).</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>trim_galore -j {threads} {params.config} -o $(dirname {output.se}) {input} &gt; {log} 2&gt;&amp;1

# now rename to proper output
if [ &quot;{params.fqsuffix}&quot; != &quot;fq&quot; ]; then
  mv &quot;$(dirname {output.se})/{wildcards.sample}_trimmed.fq.gz&quot; {output.se}
fi

# move the trimming report to qc directory
report=$(dirname {output.se})/{wildcards.sample}.{params.fqsuffix}.gz_trimming_report.txt
mv $report {output.qc}
</pre></div>
</div>
</div>
<div class="section" id="twobit">
<h2>twobit<a class="headerlink" href="#twobit" title="Permalink to this headline">¶</a></h2>
<p>Generate a 2bit file for each assembly</p>
</div>
<div class="section" id="txi-count-matrix">
<h2>txi_count_matrix<a class="headerlink" href="#txi-count-matrix" title="Permalink to this headline">¶</a></h2>
<p>Convert estimated transcript abundances to gene count estimations and merge
gene counts per assembly.</p>
<p>Also outputs a single cell experiment object similar to ARMOR
(<a class="reference external" href="https://github.com/csoneson/ARMOR">https://github.com/csoneson/ARMOR</a>).</p>
<p>Only works with Ensembl assemblies.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="schemas.html" class="btn btn-neutral float-right" title="All configurable options" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="cli.html" class="btn btn-neutral float-left" title="Command line interface" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, &amp; Simon van Heeringen.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>