

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Filling out the samples.tsv and config.yaml &mdash; seq2science  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Workflows" href="workflows.html" />
    <link rel="prev" title="Getting started" href="gettingstarted.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="home.html">Welcome</a></li>
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Filling out the samples.tsv and config.yaml</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#filling-out-the-samples-tsv">Filling out the samples.tsv</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#sample-column">sample column</a></li>
<li class="toctree-l3"><a class="reference internal" href="#assembly-column">assembly column</a></li>
<li class="toctree-l3"><a class="reference internal" href="#descriptive-name-column">descriptive_name column</a></li>
<li class="toctree-l3"><a class="reference internal" href="#working-with-technical-and-biological-replicates">working with technical and biological replicates</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#technical-replicates">Technical replicates</a></li>
<li class="toctree-l4"><a class="reference internal" href="#biological-replicates">Biological replicates</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#final-notes">Final notes</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#filling-out-the-config-yaml">Filling out the config.yaml</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
<li class="toctree-l1"><a class="reference internal" href="schemas.html">All configurable options</a></li>
<li class="toctree-l1"><a class="reference internal" href="extracontent.html">Extra resources</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Filling out the samples.tsv and config.yaml</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/fillingout.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="filling-out-the-samples-tsv-and-config-yaml">
<h1>Filling out the samples.tsv and config.yaml<a class="headerlink" href="#filling-out-the-samples-tsv-and-config-yaml" title="Permalink to this headline">¶</a></h1>
<div class="section" id="filling-out-the-samples-tsv">
<h2>Filling out the samples.tsv<a class="headerlink" href="#filling-out-the-samples-tsv" title="Permalink to this headline">¶</a></h2>
<p>Before running a workflow you will have to specify which samples you want to run the workflow on. Each workflow starts with a <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> as an example, and you should adapt it to your specific needs. As an example <code class="docutils literal notranslate"><span class="pre">samples.tsv</span></code> could look like this, but not all columns are always necessary or required:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sample    assembly    replicate    descriptive_name
GSM123    GRCh38      heart_1      heart_merged
GSM321    GRCh38      heart_1      heart_merged
GSMabc    GRCh38      heart_2      heart_not_merged
GSMxzy    danRer11    stage_8      stage_8
GSM890    danRer11    stage_9      stage_9
</pre></div>
</div>
<div class="section" id="sample-column">
<h3>sample column<a class="headerlink" href="#sample-column" title="Permalink to this headline">¶</a></h3>
<p>This column is necessary for all workflows. If you use the pipeline on public data this should be the name of the accession (e.g. GSM2837484), if you use the pipeline on local data this should be the <em>basename</em> of the file without the <em>extension</em>. For instance <code class="docutils literal notranslate"><span class="pre">/home/user/mfastqs/sample1.fastq.gz</span></code> would be sample1.</p>
</div>
<div class="section" id="assembly-column">
<h3>assembly column<a class="headerlink" href="#assembly-column" title="Permalink to this headline">¶</a></h3>
<p>This column is necessary for all workflows, except the <em>downloading samples</em> workflow. Here you simply add the name of the assembly you want your samples aligned against and the workflow will download it for you.</p>
</div>
<div class="section" id="descriptive-name-column">
<h3>descriptive_name column<a class="headerlink" href="#descriptive-name-column" title="Permalink to this headline">¶</a></h3>
<p>The descriptive_name column is used for the trackhub and multiqc report. In the trackhub your tracks will be called after the descriptive name, and in the multiqc report there will be a button to rename your samples after this column.</p>
</div>
<div class="section" id="working-with-technical-and-biological-replicates">
<h3>working with technical and biological replicates<a class="headerlink" href="#working-with-technical-and-biological-replicates" title="Permalink to this headline">¶</a></h3>
<div class="section" id="technical-replicates">
<h4>Technical replicates<a class="headerlink" href="#technical-replicates" title="Permalink to this headline">¶</a></h4>
<p>Technical replicates, or any fastq file you may wish to merge, are set using the <code class="docutils literal notranslate"><span class="pre">replicate</span></code> column in the samples.tsv file. All samples with the same name in the <code class="docutils literal notranslate"><span class="pre">replicate</span></code> column will be concatenated into one file with the replicate name.</p>
<p>Example samples.tsv utilizing replicate merging:</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>sample    assembly    replicate
GSM123    GRCh38      heart
GSMabc    GRCh38      heart
GSMxzy    danRer11    stage8
GSM890    danRer11
</pre></div>
</div>
<p>Using this file in the alignment workflow will output <em>heart.bam</em>, <em>stage8.bam</em> and <em>GSM890.bam</em>. The MultiQC will inform you of the trimming steps performed on all samples, and subsequent information of the ‘replicate’ files (of which only <em>heart</em> is merged).</p>
<div class="section" id="technical-replicates-for-scatac-seq">
<h5>Technical replicates for scATAC-seq<a class="headerlink" href="#technical-replicates-for-scatac-seq" title="Permalink to this headline">¶</a></h5>
<p>Cells processed together (e.g. sequenced from the same 384w plate) should get the same technical sample name. If your single cell experiment is from a single technical replicate (e.g. a microfluidics experiment), give all the individual cells the same technical sample name.</p>
</div>
<div class="section" id="notes">
<h5>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h5>
<p>If you are working with multiple assemblies in one workflow, replicate names have to be unique between assemblies (you will receive a warning if names overlap).</p>
<p>Replicate merging is turned on by default. It can be turned off by setting <code class="docutils literal notranslate"><span class="pre">technical_replicates</span></code> to <code class="docutils literal notranslate"><span class="pre">keep</span></code>.</p>
</div>
</div>
<div class="section" id="biological-replicates">
<h4>Biological replicates<a class="headerlink" href="#biological-replicates" title="Permalink to this headline">¶</a></h4>
<div class="section" id="rna-seq">
<h5>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">¶</a></h5>
<p>During the RNA-seq workflow, biological replicates are specified in the contrast design, as described in the <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/6.-RNA-seq">RNA-seq chapter</a>.</p>
</div>
<div class="section" id="atac-and-chip-seq">
<h5>ATAC- and ChIP-seq<a class="headerlink" href="#atac-and-chip-seq" title="Permalink to this headline">¶</a></h5>
<p>During ATAC- and ChIP-seq workflows, peak calling can be performed per biological condition, depending on your configuration setting. Biological conditions are determined by the <code class="docutils literal notranslate"><span class="pre">condition</span></code> column in the samples.tsv file. How these samples are handled is specified by configuration variable <code class="docutils literal notranslate"><span class="pre">biological_replicates</span></code>.</p>
<p>If you merge technical replicates <em>and</em> combine biological replicates, the technical replicates are merged first.</p>
</div>
<div class="section" id="keep">
<h5>Keep<a class="headerlink" href="#keep" title="Permalink to this headline">¶</a></h5>
<p>The default setting. All samples/merged replicates are analyzed individually. The <code class="docutils literal notranslate"><span class="pre">condition</span></code> column is ignored.</p>
</div>
<div class="section" id="irreproducible-discovery-rate-idr">
<h5>Irreproducible Discovery Rate (IDR)<a class="headerlink" href="#irreproducible-discovery-rate-idr" title="Permalink to this headline">¶</a></h5>
<p>One of the more common methods to combine biological replicates is by the irreproducible discovery rate (<a class="reference external" href="https://github.com/kundajelab/idr">idr</a>). Shortly; idr sorts all the peaks of two replicates separately on their significance. Since true peaks should be very significant for both replicates these peaks will be one of the highest sorted peaks. As peaks get less and less true (and thus their significance) their ordering also becomes more random between the samples. The idr method then only keeps the peak that overlap <em>nonrandomly</em> between the samples. The idr method only works for two replicates, so can not be used when you have more than 2 (<code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">==</span> <span class="pre">2</span></code>).</p>
</div>
<div class="section" id="fisher-s-method">
<h5>Fisher’s method<a class="headerlink" href="#fisher-s-method" title="Permalink to this headline">¶</a></h5>
<p><a class="reference external" href="https://en.wikipedia.org/wiki/Fisher%27s_method">Fisher’s method</a> simply is a method to ‘combine’ multiple p-values. This method is built-in for genrich and macs2, and allows any number of replicates (<code class="docutils literal notranslate"><span class="pre">n</span> <span class="pre">&gt;=</span> <span class="pre">2</span></code>). However a ‘disadvantage’ of this method is that it assumes the p-values a method generates are actual p-values (and not a general indication of significance). MACS2 is slightly notorious for that its p-values are not true p-values. Genrich claims by fitting on a log-normal distribution that their p-values are true. Even though fisher’s method might not be ideal, it’s the only option you have (we provide) within this pipeline for more than 2 replicates.</p>
</div>
</div>
</div>
<div class="section" id="final-notes">
<h3>Final notes<a class="headerlink" href="#final-notes" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Make sure that the samples.tsv is a tab separated values file when running the pipeline.</p></li>
<li><p>Feel free to delete or add columns to your liking.</p></li>
</ul>
</div>
</div>
<div class="section" id="filling-out-the-config-yaml">
<h2>Filling out the config.yaml<a class="headerlink" href="#filling-out-the-config-yaml" title="Permalink to this headline">¶</a></h2>
<p>Every workflow has many configurable options, and can be set in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file. In each <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> we highlighted a couple options that we think are relevant for that specific workflow, and set (we think) <strong>reasonable default</strong> values.</p>
<p>When a workflow starts it prints the complete configuration, and (almost) all these values can be added in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> and changed to your liking. You can see the complete set of configurable options <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/tree/master/schemas/config">here</a>.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="workflows.html" class="btn btn-neutral float-right" title="Workflows" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="gettingstarted.html" class="btn btn-neutral float-left" title="Getting started" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, &amp; Simon van Heeringen.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>