

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>All configurable options &mdash; seq2science  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Extra resources" href="extracontent.html" />
    <link rel="prev" title="Using the results" href="results.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="home.html">Welcome</a></li>
<li class="toctree-l1"><a class="reference internal" href="gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="fillingout.html">Filling out the samples.tsv and config.yaml</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Using the results</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">All configurable options</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#general">general</a></li>
<li class="toctree-l2"><a class="reference internal" href="#peak-calling-chip-atac">peak calling (ChIP &amp; ATAC)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#gene-expression-rna-seq">gene expression (RNA-seq)</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>All configurable options</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/schemas.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="all-configurable-options">
<h1>All configurable options<a class="headerlink" href="#all-configurable-options" title="Permalink to this headline">¶</a></h1>
<p>TODO
This is an automatically generated …
These are all configurable options, loosely spread around topics ….</p>
<div class="section" id="general">
<h2>general<a class="headerlink" href="#general" title="Permalink to this headline">¶</a></h2>
<p>samples:
description: tab-separated file specifying the samples to run through the workflow
default: samples.tsv
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">rule_dir:</span>
<span class="c">description: directory where to find the workflow rules</span>
<span class="c">default: ../../rules</span>
<span class="c">type: string</span>
</pre></div>
</div>
<p>result_dir:
description: where to store the pipeline’s results (by default)
default: ./
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">genome_dir:</span>
<span class="c">description: where to store the pipeline&#39;s genome assemblies</span>
<span class="c">default: ./</span>
<span class="c">type: string</span>
</pre></div>
</div>
<p>log_dir:
description: directory where to store the logs (defaults inside the result dir)
default: log
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">benchmark_dir:</span>
<span class="c">description: directory where to store the logs (defaults inside the result dir)</span>
<span class="c">default: benchmark</span>
<span class="c">type: string</span>
</pre></div>
</div>
<p>fqext1:
description: filename suffix when handling paired-end data, describing the forward read
default: R1
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">fqext2:</span>
<span class="c">description: filename suffix when handling paired-end data, describing the reverse read</span>
<span class="c">default: R2</span>
<span class="c">type: string</span>
</pre></div>
</div>
<p>fqsuffix:
description: file descriptor for fastq files (often fastq or fq)
default: fastq
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">cpulimit:</span>
<span class="c">description: whether or not to make use of cpulimit to enfoce max thread usage for some rules</span>
<span class="c">default: True</span>
<span class="c">type: boolean</span>
</pre></div>
</div>
<p>email:
description: email to reach you after pipeline finished, required for trackhub and multiqc report
default: <a class="reference external" href="mailto:none&#37;&#52;&#48;provided&#46;com">none<span>&#64;</span>provided<span>&#46;</span>com</a>
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>## alignment sth
bam_sorter:
description: which sorting software to use
type: object
properties:
    description: sort order of aligned reads
    default: coordinate
    enum: [&#39;queryname&#39;, &#39;coordinate&#39;]
    type: string
default:
    samtools: coordinate
</pre></div>
</div>
<p>cram_no_bam:
description: whether to store cram instead of bam
type: boolean
default: False</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>## alignment
trimmed_dir:
description: directory where to store the workflow trimmed fastqs (defaults inside the result dir)
default: fastq_trimmed
type: string
</pre></div>
</div>
<p>qc_dir:
description: directory where to store the workflow quality controls (defaults inside the result dir)
default: qc
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">dedup_dir:</span>
<span class="c">description: directory where to store the workflow deduplicated alignment (defaults inside the result dir)</span>
<span class="c">default: dedup</span>
<span class="c">type: string</span>
</pre></div>
</div>
<p>trim_galore:
description: trim_galore settings (note that the adapter is autodetected, so you do not have to set it here)
default: –quality 10 –length 20 –consider_already_trimmed 10
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">technical_replicates:</span>
<span class="c">description: concatenates samples depending on the names given in the &#39;replicates&#39; column of samples.tsv</span>
<span class="c">default: merge</span>
<span class="c">enum: [&#39;merge&#39;, &#39;keep&#39;]</span>
</pre></div>
</div>
<p>biological_replicates:
description: which method to use to combine replicates (fisher (n &gt;=2), idr (n==2), or keep)
default: keep
enum: [‘fisher’, ‘idr’, ‘keep’]
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">aligner:</span>
<span class="c">description: which aligner to use</span>
<span class="c">properties:</span>
<span class="c">    bowtie2:</span>
<span class="c">        properties:</span>
<span class="c">            index:</span>
<span class="c">                default:</span>
<span class="c">            align:</span>
<span class="c">                default:</span>
<span class="c">        default:</span>
<span class="c">            index:</span>
<span class="c">            align:</span>
<span class="c">    bwa:</span>
<span class="c">        properties:</span>
<span class="c">            index:</span>
<span class="c">                default: -a bwtsw</span>
<span class="c">            align:</span>
<span class="c">                default: -M</span>
<span class="c">        default:</span>
<span class="c">            index: -a bwtsw</span>
<span class="c">            align: -M</span>
<span class="c">    hisat2:</span>
<span class="c">        properties:</span>
<span class="c">            index:</span>
<span class="c">                default:</span>
<span class="c">            align:</span>
<span class="c">                default:</span>
<span class="c">        default:</span>
<span class="c">            index:</span>
<span class="c">            align:</span>
<span class="c">    star:</span>
<span class="c">        properties:</span>
<span class="c">            index:</span>
<span class="c">                default: --limitGenomeGenerateRAM 37000000000 --genomeSAsparseD 1</span>
<span class="c">            align:</span>
<span class="c">                default:</span>
<span class="c">        default:</span>
<span class="c">            index: --limitGenomeGenerateRAM 37000000000 --genomeSAsparseD 1</span>
<span class="c">            align:</span>
<span class="c">default:</span>
<span class="c">    bwa:</span>
<span class="c">        index: -a bwtsw</span>
<span class="c">        align:</span>
</pre></div>
</div>
<p>samtools_index:
description: samtools index settings
default:
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">markduplicates:</span>
<span class="c">description: the parameters of picard markduplicates</span>
<span class="c">default: REMOVE_DUPLICATES=false -Xms4G -Xmx6G MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=999</span>
<span class="c">type: string</span>
</pre></div>
</div>
<p>min_mapping_quality:
default: 0
type: integer</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">tn5_shift:</span>
<span class="c">default: False</span>
<span class="c">type: boolean</span>
</pre></div>
</div>
<p>remove_blacklist:
default: False</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">remove_mito:</span>
<span class="c">default: False</span>
</pre></div>
</div>
<p>only_primary_align:
default: False</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">create_qc_report:</span>
<span class="c">default: True</span>
<span class="c">type: boolean</span>
</pre></div>
</div>
</div>
<div class="section" id="peak-calling-chip-atac">
<h2>peak calling (ChIP &amp; ATAC)<a class="headerlink" href="#peak-calling-chip-atac" title="Permalink to this headline">¶</a></h2>
<p>peak_caller:
type: object
properties:</p>
<blockquote>
<div><dl class="simple">
<dt>genrich:</dt><dd><p>type: string</p>
</dd>
<dt>hmmratac:</dt><dd><p>type: string</p>
</dd>
<dt>macs2:</dt><dd><p>type: string</p>
</dd>
</dl>
</div></blockquote>
<p>minProperties: 1
additionalProperties: False
default:</p>
<blockquote>
<div><p>macs2: –shift -100 –extsize 200 –nomodel –keep-dup 1 –buffer-size 10000</p>
</div></blockquote>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">macs2_keep_mates:</span>
<span class="c">description: which method to use to combine replicates (idr or fisher)</span>
<span class="c">default: False</span>
<span class="c">type: boolean</span>
</pre></div>
</div>
</div>
<div class="section" id="gene-expression-rna-seq">
<h2>gene expression (RNA-seq)<a class="headerlink" href="#gene-expression-rna-seq" title="Permalink to this headline">¶</a></h2>
<p>counts_dir:
description: directory where to store the gene counts for DGE analysis (defaults inside the result dir)
default: gene_counts
type: string</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">dge_dir:</span>
<span class="c">description: directory where to store the workflow differential gene analysis output (defaults inside the result dir)</span>
<span class="c">default: deseq2</span>
<span class="c">type: string</span>
</pre></div>
</div>
<p>quantifier:
description: which quantifier to use
properties:</p>
<blockquote>
<div><dl class="simple">
<dt>star:</dt><dd><dl class="simple">
<dt>properties:</dt><dd><dl class="simple">
<dt>quantify:</dt><dd><p>default:</p>
</dd>
</dl>
</dd>
<dt>default:</dt><dd><p>quantify:</p>
</dd>
</dl>
</dd>
<dt>salmon:</dt><dd><dl class="simple">
<dt>properties:</dt><dd><dl class="simple">
<dt>quantifier_index:</dt><dd><p>default: -k 31</p>
</dd>
<dt>quantify:</dt><dd><p>default: –seqBias –gcBias –validateMappings –recoverOrphans</p>
</dd>
<dt>decoy_aware_index:</dt><dd><p>default: False</p>
</dd>
</dl>
</dd>
<dt>default:</dt><dd><p>quantifier_index: -k 31
quantify: –seqBias –gcBias –validateMappings –recoverOrphans
decoy_aware_index: False</p>
</dd>
</dl>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>default:</dt><dd><dl class="simple">
<dt>star:</dt><dd><p>quantify:</p>
</dd>
</dl>
</dd>
</dl>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">tximeta:</span>
<span class="c">description: tximeta is used to convert salmon output (estimated transcript abundances) to a gene counts matrix. Currently, these values can be ignored.</span>
<span class="c">properties:</span>
<span class="c">    source:</span>
<span class="c">        default: Ensembl</span>
<span class="c">    organism:</span>
<span class="c">        default: tximeta_file</span>
<span class="c">    release:</span>
<span class="c">        default: 42</span>
<span class="c">default:</span>
<span class="c">    soure: Ensembl</span>
<span class="c">    organism: tximeta_file</span>
<span class="c">    release: 42</span>
</pre></div>
</div>
<p>diffexp:
description: which differential expression analysis to use (DESeq2 or None, default is None). Will not run without a contrast.
type: object
properties:</p>
<blockquote>
<div><dl class="simple">
<dt>deseq2:</dt><dd><dl class="simple">
<dt>properties:</dt><dd><dl class="simple">
<dt>DE_params:</dt><dd><dl class="simple">
<dt>default:</dt><dd><p>multiple_testing_procedure: BH
alpha_value: 0.05
shrinkage_estimator: apeglm</p>
</dd>
</dl>
</dd>
<dt>contrasts:</dt><dd><p>default:</p>
</dd>
</dl>
</dd>
<dt>default:</dt><dd><p>contrasts:</p>
</dd>
</dl>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>default:</dt><dd><dl class="simple">
<dt>None:</dt><dd><p>contrasts:</p>
</dd>
</dl>
</dd>
</dl>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span>## trackhub
create_trackhub:
description: skip generation of bams (except for workflow alignment), bigwigs and a trackhub
default: True
enum: [True, False]
</pre></div>
</div>
<p>deeptools:
description: option flags for bam to bigwig conversions with deeptools bamCoverage
default: –samFlagExclude 1024</p>
<div class="highlight-guess notranslate"><div class="highlight"><pre><span></span><span class="c">filter_bam_by_strand:</span>
<span class="c">description: split bigwigs by strand?</span>
<span class="c">default: True</span>
<span class="c">enum: [True, False]</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="extracontent.html" class="btn btn-neutral float-right" title="Extra resources" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="results.html" class="btn btn-neutral float-left" title="Using the results" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, &amp; Simon van Heeringen.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>