

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>RNA-seq &mdash; seq2science  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="ChIP-seq" href="chip_seq.html" />
    <link rel="prev" title="ATAC-seq" href="atac_seq.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> seq2science
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../gettingstarted.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../fillingout.html">Filling out the samples.tsv and config.yaml</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../workflows.html">Workflows</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="download_fastq.html">Downloading fastqs</a></li>
<li class="toctree-l2"><a class="reference internal" href="alignment.html">Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="atac_seq.html">ATAC-seq</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">RNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pipeline-steps">Pipeline steps</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#generate-count-matrices">Generate count matrices</a></li>
<li class="toctree-l4"><a class="reference internal" href="#trackhub">Trackhub</a></li>
<li class="toctree-l4"><a class="reference internal" href="#differential-gene-expression-analysis">Differential gene expression analysis</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#contrast-designs">Contrast designs</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#conditions">Conditions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#batches">Batches</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#differential-gene-expression-analysis-output">Differential gene expression analysis output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="chip_seq.html">ChIP-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="scatac_seq.html">scATAC-seq</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Using the results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../schemas.html">All configurable options</a></li>
<li class="toctree-l1"><a class="reference internal" href="../extracontent.html">Extra resources</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">Frequently Asked Questions (FAQ)</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">seq2science</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../workflows.html">Workflows</a> &raquo;</li>
        
      <li>RNA-seq</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/vanheeringen-lab/seq2science/blob/master/docs/content/workflows/rna_seq.md" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="rna-seq">
<h1>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">¶</a></h1>
<p>Running an RNA-seq pipeline has never been easier! See our <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/tree/master/workflows/rna_seq">RNA-seq</a> workflow.</p>
<div class="section" id="pipeline-steps">
<h2>Pipeline steps<a class="headerlink" href="#pipeline-steps" title="Permalink to this headline">¶</a></h2>
<div class="section" id="generate-count-matrices">
<h3>Generate count matrices<a class="headerlink" href="#generate-count-matrices" title="Permalink to this headline">¶</a></h3>
<p>In addition to aligning to the genome, <a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> can output a reads per gene table. These counts are required for differential expression analysis.</p>
<p>If alignment is no priority, either STAR or <a class="reference external" href="https://github.com/COMBINE-lab/salmon">Salmon</a> can be used to only quantify the transcripts. This option is used when <code class="docutils literal notranslate"><span class="pre">generate_trackhub</span></code> is set to <code class="docutils literal notranslate"><span class="pre">False</span></code> (the default for this workflow).</p>
</div>
<div class="section" id="trackhub">
<h3>Trackhub<a class="headerlink" href="#trackhub" title="Permalink to this headline">¶</a></h3>
<p>A trackhub can be generated for this workflow. See the <a class="reference external" href="https://github.com/vanheeringen-lab/snakemake-workflows/wiki/Trackhub">Trackhub page</a> for more information!</p>
</div>
<div class="section" id="differential-gene-expression-analysis">
<h3>Differential gene expression analysis<a class="headerlink" href="#differential-gene-expression-analysis" title="Permalink to this headline">¶</a></h3>
<p>Differential expression analysis can automatically be performed using <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">DESeq2</a>. Simply add one or more contrast design(s) in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file. Examples are given below. Additionally, the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> contains a commented-out set of examples.</p>
<p>Note: (additional) design contrasts can be added at any time. After completing the workflow, rerunning Snakemake with new contrasts will only perform these analyses.</p>
<div class="section" id="id2">
<h4>DESeq2<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h4>
<p>DESeq2 automatically performs library bias correction when loading your data. Batch correction is performed if a batch effect is included in the design. After calculating differentially expressed genes, a multiple testing procedure is applied. By default this is the Benjamini-Hochberg procedure, although Independent Hypothesis Weighing can also be applied. Expression counts are log transformed (by default using the apeglm method). Finally, a list of all genes is saved to file, with analysis results for expressed genes.</p>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="contrast-designs">
<h2>Contrast designs<a class="headerlink" href="#contrast-designs" title="Permalink to this headline">¶</a></h2>
<p>The following section will guide you through the process of creating a DESeq2 contrast using only samples.tsv and the config.yaml.</p>
<p>A contrast contains a condition, which is the variable you wish to analyse, and which contains 2 or more states. Optionally, a contrast can contain a batch effect, which you want to correct for.</p>
<div class="section" id="conditions">
<h3>Conditions<a class="headerlink" href="#conditions" title="Permalink to this headline">¶</a></h3>
<p>In samples.tsv, add a column for every property of your samples.
Examples:</p>
<ul class="simple">
<li><p>a column named ‘conditions’ with values ‘control’ and ‘treated’.</p></li>
<li><p>a column named ‘conditions’ with values ‘control’, ‘treatmentA’ and ‘treatmentB’.</p></li>
<li><p>a column named ‘stages’ with values 1, 2 and 3.</p></li>
</ul>
<p>Next, in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> add one or more contrasts.</p>
<p>In order to compare two groups (e.a. control vs treated) in a column with only two groups, the contrast condition is the column name (<code class="docutils literal notranslate"><span class="pre">conditions</span></code>) (the next notation is allowed as well).</p>
<p>In order to compare two groups for a larger list (e.a. stage 1 vs stage 3, but not stage 2), the contrast condition is the column name, followed by the groups (<code class="docutils literal notranslate"><span class="pre">stages_3_1</span></code>).</p>
<p>To compare all groups against a reference (e.a. control vs both treatments A and B), the contrast condition is the column name, followed by the reference level (<code class="docutils literal notranslate"><span class="pre">condition_control</span></code>). Alternatively, group ‘all’ may also be added (<code class="docutils literal notranslate"><span class="pre">condition_all_control</span></code>).</p>
</div>
<div class="section" id="batches">
<h3>Batches<a class="headerlink" href="#batches" title="Permalink to this headline">¶</a></h3>
<p>If your data has batch effects, these can be accounted for. Note that this can only be done if the batch effects are spread over different conditions.</p>
<p>Similar to conditions, add a column to samples.tsv and note the batch of each sample:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>sample</p></th>
<th class="head"><p>sequencing_month</p></th>
<th class="head"><p>conditions</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>sample_1</p></td>
<td><p>jan</p></td>
<td><p>control</p></td>
</tr>
<tr class="row-odd"><td><p>sample_2</p></td>
<td><p>feb</p></td>
<td><p>control</p></td>
</tr>
<tr class="row-even"><td><p>sample_3</p></td>
<td><p>jan</p></td>
<td><p>treatment</p></td>
</tr>
<tr class="row-odd"><td><p>sample_4</p></td>
<td><p>feb</p></td>
<td><p>treatment</p></td>
</tr>
</tbody>
</table>
<p>In this case, the sequencing_month may have caused a batch effect. Since the (potential) batch effect is spread over the test conditions (control vs treatment) it can be taken into account during DE analysis. The contrast design to do this would be <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_treatment_control</span></code></p>
<div class="section" id="more-examples">
<h4>More examples<a class="headerlink" href="#more-examples" title="Permalink to this headline">¶</a></h4>
<p>The example design <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_treatment_control</span></code> may also be written as <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions</span></code> or <code class="docutils literal notranslate"><span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_all_control</span></code>.</p>
<p>For consistency with regular DESeq2 contrast designs, designs may start with <code class="docutils literal notranslate"><span class="pre">~</span></code>, example: <code class="docutils literal notranslate"><span class="pre">~</span> <span class="pre">sequencing_month</span> <span class="pre">+</span> <span class="pre">conditions_treatment_control</span></code>.</p>
<p>Our gremlins will carefully unpack all of your input and pass the unique designs to DESeq2.</p>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="differential-gene-expression-analysis-output">
<h2>Differential gene expression analysis output<a class="headerlink" href="#differential-gene-expression-analysis-output" title="Permalink to this headline">¶</a></h2>
<p>Each contrast will return a table with all genes found during alignment.</p>
<ul class="simple">
<li><p>The column <code class="docutils literal notranslate"><span class="pre">padj</span></code> contains the adjusted p-values after multiple testing. These should be used extract DE genes.</p></li>
<li><p>The column <code class="docutils literal notranslate"><span class="pre">log2FoldChange</span></code> shows the fold change of each gene between the two conditions. The reference condition is the one <em>last mentioned</em> in the contrast design, so use <code class="docutils literal notranslate"><span class="pre">conditions_treatment_control</span></code>. If you use <code class="docutils literal notranslate"><span class="pre">conditions_control_treatment</span></code> the fold change is <em>inverted</em>.</p></li>
<li><p>Several other columns were kept for sake of completion, such as column <code class="docutils literal notranslate"><span class="pre">pvalue</span></code>, which contains non-adjusted p-values and <em>should not be used</em>.</p></li>
</ul>
<p>Additional outputs include a blind clustering (which may reveal switched sample names), and per contrast an MA plot and a PCA plot (to accompany the differentially expressed genes table).</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="chip_seq.html" class="btn btn-neutral float-right" title="ChIP-seq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="atac_seq.html" class="btn btn-neutral float-left" title="ATAC-seq" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright Maarten van der Sande, Siebren Frölich, Jos Smits, &amp; Simon van Heeringen.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>