from snakemake.utils import validate
configfile: "config.yaml"
validate(config, schema=f"../../schemas/config.schema.yaml", set_default=True)


# complement the configuration and do onstart/onexit things
schema = 'alignment.yaml'
include: f"{config['rule_dir']}/configuration.smk"


# load the subworkflow download_fastq
subworkflow download_fastq:
    workdir:
        "../download_fastq"
    configfile:
        "config.yaml"


# load the remaining relevant rules
include: f"{config['rule_dir']}/get_genome.smk"
include: f"{config['rule_dir']}/alignment.smk"
include: f"{config['rule_dir']}/trim_auto.smk"


rule all:
    """
    align each sample against its assembly
    """
    input:
        [expand(f"{{result_dir}}/{{bwa_dir}}/{sample}-{samples.loc[sample]['assembly']}.bam", **config) for sample in samples.index],
        [expand(f"{{result_dir}}/{{bwa_dir}}/{sample}-{samples.loc[sample]['assembly']}.bai", **config) for sample in samples.index]
